{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/hariszaf/metabolic_toy_model/blob/main/Antony2025/gapfillingGSMMs.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "u2wfZmrxxAaJ"
      },
      "source": [
        "# **Gapfilling Genome-Scale Metabolic Models with DNNGIOR**"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install cobra\n",
        "!pip install dnngior --no-deps"
      ],
      "metadata": {
        "id": "dpzmKk5RcwjT",
        "outputId": "87bc14e1-142a-4c19-a218-a72dcd995001",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: cobra in /usr/local/lib/python3.11/dist-packages (0.29.1)\n",
            "Requirement already satisfied: appdirs~=1.4 in /usr/local/lib/python3.11/dist-packages (from cobra) (1.4.4)\n",
            "Requirement already satisfied: depinfo~=2.2 in /usr/local/lib/python3.11/dist-packages (from cobra) (2.2.0)\n",
            "Requirement already satisfied: diskcache~=5.0 in /usr/local/lib/python3.11/dist-packages (from cobra) (5.6.3)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.11/dist-packages (from cobra) (1.0.0)\n",
            "Requirement already satisfied: httpx~=0.24 in /usr/local/lib/python3.11/dist-packages (from cobra) (0.28.1)\n",
            "Requirement already satisfied: importlib-resources in /usr/local/lib/python3.11/dist-packages (from cobra) (6.5.2)\n",
            "Requirement already satisfied: numpy>=1.13 in /usr/local/lib/python3.11/dist-packages (from cobra) (1.26.4)\n",
            "Requirement already satisfied: optlang~=1.8 in /usr/local/lib/python3.11/dist-packages (from cobra) (1.8.3)\n",
            "Requirement already satisfied: pandas<3.0,>=1.0 in /usr/local/lib/python3.11/dist-packages (from cobra) (2.2.2)\n",
            "Requirement already satisfied: pydantic>=1.6 in /usr/local/lib/python3.11/dist-packages (from cobra) (2.10.6)\n",
            "Requirement already satisfied: python-libsbml~=5.19 in /usr/local/lib/python3.11/dist-packages (from cobra) (5.20.4)\n",
            "Requirement already satisfied: rich>=8.0 in /usr/local/lib/python3.11/dist-packages (from cobra) (13.9.4)\n",
            "Requirement already satisfied: ruamel.yaml~=0.16 in /usr/local/lib/python3.11/dist-packages (from cobra) (0.18.10)\n",
            "Requirement already satisfied: swiglpk in /usr/local/lib/python3.11/dist-packages (from cobra) (5.0.12)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (3.7.1)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (2025.1.31)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (1.0.7)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (3.10)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.11/dist-packages (from httpcore==1.*->httpx~=0.24->cobra) (0.14.0)\n",
            "Requirement already satisfied: sympy>=1.12.0 in /usr/local/lib/python3.11/dist-packages (from optlang~=1.8->cobra) (1.13.1)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0,>=1.0->cobra) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0,>=1.0->cobra) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0,>=1.0->cobra) (2025.1)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.2 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra) (2.27.2)\n",
            "Requirement already satisfied: typing-extensions>=4.12.2 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra) (4.12.2)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich>=8.0->cobra) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich>=8.0->cobra) (2.18.0)\n",
            "Requirement already satisfied: ruamel.yaml.clib>=0.2.7 in /usr/local/lib/python3.11/dist-packages (from ruamel.yaml~=0.16->cobra) (0.2.12)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich>=8.0->cobra) (0.1.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas<3.0,>=1.0->cobra) (1.17.0)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy>=1.12.0->optlang~=1.8->cobra) (1.3.0)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio->httpx~=0.24->cobra) (1.3.1)\n",
            "Requirement already satisfied: dnngior in /usr/local/lib/python3.11/dist-packages (0.0.5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# @title\n",
        "import os\n",
        "def create_gurobi_license():\n",
        "    license_content = (\n",
        "        \"# Gurobi WLS license file\\n\"\n",
        "        \"# Your credentials are private and should not be shared or copied to public repositories.\\n\"\n",
        "        \"# Visit https://license.gurobi.com/manager/doc/overview for more information.\\n\"\n",
        "        \"WLSACCESSID=1fedf73b-9471-4da8-bdc7-2aaacf2e30f3\\n\"\n",
        "        \"WLSSECRET=3bc7d209-a4ec-4195-98be-4b254f181512\\n\"\n",
        "        \"LICENSEID=940603\"\n",
        "    )\n",
        "    with open(\"/content/licenses/gurobi.lic\", \"w\") as f:\n",
        "        f.write(license_content)\n",
        "    print(\"License file created at /content/licenses/gurobi.lic\")\n",
        "\n",
        "\n",
        "\n",
        "# Create directory for the license\n",
        "os.makedirs(\"/content/licenses\", exist_ok=True)\n",
        "\n",
        "# Generate the license file\n",
        "create_gurobi_license()\n",
        "\n",
        "#add to path\n",
        "os.environ['GRB_LICENSE_FILE'] = '/content/licenses/gurobi.lic'\n",
        "\n",
        "!pip install gurobipy"
      ],
      "metadata": {
        "id": "_sqCJuwTdLKG",
        "outputId": "7deec621-bf67-4423-a91b-600f585a6447",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "License file created at /content/licenses/gurobi.lic\n",
            "Requirement already satisfied: gurobipy in /usr/local/lib/python3.11/dist-packages (12.0.1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')"
      ],
      "metadata": {
        "id": "1S-0-ShJcWnU",
        "outputId": "0a62a028-260f-4d44-b261-88098aa5683f",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount(\"/content/drive/\", force_remount=True).\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "cd /content/drive/MyDrive/Antony"
      ],
      "metadata": {
        "id": "f1ZLti_gcmW6",
        "outputId": "2c001d7f-d1ab-4e4b-a9f1-61205f7ca0ec",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/Antony\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cgA4LKFjcSAP"
      },
      "source": [
        "### Gapfilling a model with a complete medium"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HZVTBTTVcSAQ"
      },
      "source": [
        "Loading depencies"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "-Ajq7ymow_oo",
        "outputId": "922dd10e-19a5-4673-ad8e-e3b08fd02419",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Set parameter WLSAccessID\n",
            "Set parameter WLSSecret\n",
            "Set parameter LicenseID to value 940603\n",
            "Academic license 940603 - for non-commercial use only - registered to da___@gmail.com\n",
            "WARNING: To enable the NN_Trainer script, you need to install tensorflow <https://www.tensorflow.org/install>\n",
            "The rest of dnngior features can be used without it.\n"
          ]
        }
      ],
      "source": [
        "from cobra.io import read_sbml_model\n",
        "from dnngior.gapfill_class import *\n",
        "from dnngior.variables import *\n",
        "from dnngior.NN_Predictor import NN\n",
        "import pandas as pd"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uX_HQACicSAS"
      },
      "source": [
        "Let's get the Bifidobacterium model we have been working with last time"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "UpgM7rDtcSAS",
        "outputId": "49c6d253-3e61-4b31-971b-432dd3c8edd7",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 239
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<cobra.summary.model_summary.ModelSummary at 0x79b1351a0d50>"
            ],
            "text/html": [
              "<h3>Objective</h3><p>1.0 bio1 = 0.0</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "source": [
        "path_to_draft_model = \"Bifidobacterium adolescentis_atcc_15703.sbml\"\n",
        "draft_model = read_sbml_model(path_to_draft_model)\n",
        "draft_model.summary()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qlfjtRXucSAT"
      },
      "source": [
        "If we want to gapfill on a complete medium and with the default NN, its actually quite simple. We can use the gapfill class of DNNGIOR giving the path to our draft model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "F5--WfHqcSAU",
        "outputId": "18583c97-f3f1-4cee-b7d9-8d940f4c4501",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#reactions not found in NN-keys:  65 / 726\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  79156 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39578 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19789 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9894 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4947 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2473 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1236 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  618 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  309 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  154 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  77 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.061653.\n",
            "Read LP format model from file /tmp/tmp83abp6d5.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 785 rows, 1614 columns, 7118 nonzeros\n",
            "NN gapfilling added 81 new reactions\n",
            "The NN gapfilled model, comes with 807 reactions and 785 metabolites\n"
          ]
        }
      ],
      "source": [
        "gapfill_complete_medium = Gapfill(draftModel = path_to_draft_model, objectiveName = 'bio1')\n",
        "gf_model = gapfill_complete_medium.gapfilledModel"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "6_aqx0UjcSAU",
        "outputId": "1cded2a9-437e-4513-9eab-42b1d03b255e",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<cobra.summary.model_summary.ModelSummary at 0x79b12a298190>"
            ],
            "text/html": [
              "<h3>Objective</h3><p>1.0 bio1 = 61.65259806105101</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>cpd00028_e0</td>\n",
              "      <td>EX_cpd00028_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>34</td>\n",
              "      <td>0.07%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00030_e0</td>\n",
              "      <td>EX_cpd00030_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00034_e0</td>\n",
              "      <td>EX_cpd00034_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00051_e0</td>\n",
              "      <td>EX_cpd00051_e0</td>\n",
              "      <td>11.91</td>\n",
              "      <td>6</td>\n",
              "      <td>0.34%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00053_e0</td>\n",
              "      <td>EX_cpd00053_e0</td>\n",
              "      <td>358.2</td>\n",
              "      <td>5</td>\n",
              "      <td>8.58%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00058_e0</td>\n",
              "      <td>EX_cpd00058_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00060_e0</td>\n",
              "      <td>EX_cpd00060_e0</td>\n",
              "      <td>7.219</td>\n",
              "      <td>5</td>\n",
              "      <td>0.17%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00063_e0</td>\n",
              "      <td>EX_cpd00063_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00065_e0</td>\n",
              "      <td>EX_cpd00065_e0</td>\n",
              "      <td>3.355</td>\n",
              "      <td>11</td>\n",
              "      <td>0.18%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00066_e0</td>\n",
              "      <td>EX_cpd00066_e0</td>\n",
              "      <td>10.85</td>\n",
              "      <td>9</td>\n",
              "      <td>0.47%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00067_e0</td>\n",
              "      <td>EX_cpd00067_e0</td>\n",
              "      <td>73.83</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00069_e0</td>\n",
              "      <td>EX_cpd00069_e0</td>\n",
              "      <td>6.836</td>\n",
              "      <td>9</td>\n",
              "      <td>0.29%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00082_e0</td>\n",
              "      <td>EX_cpd00082_e0</td>\n",
              "      <td>998.5</td>\n",
              "      <td>6</td>\n",
              "      <td>28.70%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00099_e0</td>\n",
              "      <td>EX_cpd00099_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00104_e0</td>\n",
              "      <td>EX_cpd00104_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>10</td>\n",
              "      <td>0.02%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00107_e0</td>\n",
              "      <td>EX_cpd00107_e0</td>\n",
              "      <td>21.39</td>\n",
              "      <td>6</td>\n",
              "      <td>0.61%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00156_e0</td>\n",
              "      <td>EX_cpd00156_e0</td>\n",
              "      <td>18.94</td>\n",
              "      <td>5</td>\n",
              "      <td>0.45%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00166_e0</td>\n",
              "      <td>EX_cpd00166_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>72</td>\n",
              "      <td>0.15%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00205_e0</td>\n",
              "      <td>EX_cpd00205_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00220_e0</td>\n",
              "      <td>EX_cpd00220_e0</td>\n",
              "      <td>0.8867</td>\n",
              "      <td>17</td>\n",
              "      <td>0.07%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00254_e0</td>\n",
              "      <td>EX_cpd00254_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00264_e0</td>\n",
              "      <td>EX_cpd00264_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>7</td>\n",
              "      <td>0.01%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00276_e0</td>\n",
              "      <td>EX_cpd00276_e0</td>\n",
              "      <td>1000</td>\n",
              "      <td>6</td>\n",
              "      <td>28.75%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00305_e0</td>\n",
              "      <td>EX_cpd00305_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>12</td>\n",
              "      <td>0.03%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00322_e0</td>\n",
              "      <td>EX_cpd00322_e0</td>\n",
              "      <td>16.65</td>\n",
              "      <td>6</td>\n",
              "      <td>0.48%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd01017_e0</td>\n",
              "      <td>EX_cpd01017_e0</td>\n",
              "      <td>3.638</td>\n",
              "      <td>5</td>\n",
              "      <td>0.09%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd01329_e0</td>\n",
              "      <td>EX_cpd01329_e0</td>\n",
              "      <td>166.9</td>\n",
              "      <td>36</td>\n",
              "      <td>28.78%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd02229_e0</td>\n",
              "      <td>EX_cpd02229_e0</td>\n",
              "      <td>0.1907</td>\n",
              "      <td>55</td>\n",
              "      <td>0.05%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd10515_e0</td>\n",
              "      <td>EX_cpd10515_e0</td>\n",
              "      <td>0.8867</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd10516_e0</td>\n",
              "      <td>EX_cpd10516_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd11217_e0</td>\n",
              "      <td>EX_cpd11217_e0</td>\n",
              "      <td>0.1612</td>\n",
              "      <td>32</td>\n",
              "      <td>0.02%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd11606_e0</td>\n",
              "      <td>EX_cpd11606_e0</td>\n",
              "      <td>0.6554</td>\n",
              "      <td>46</td>\n",
              "      <td>0.14%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15311_e0</td>\n",
              "      <td>EX_cpd15311_e0</td>\n",
              "      <td>0.763</td>\n",
              "      <td>39</td>\n",
              "      <td>0.14%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15511_e0</td>\n",
              "      <td>EX_cpd15511_e0</td>\n",
              "      <td>0.09537</td>\n",
              "      <td>80</td>\n",
              "      <td>0.04%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15533_e0</td>\n",
              "      <td>EX_cpd15533_e0</td>\n",
              "      <td>1.546</td>\n",
              "      <td>41</td>\n",
              "      <td>0.30%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15540_e0</td>\n",
              "      <td>EX_cpd15540_e0</td>\n",
              "      <td>4.639</td>\n",
              "      <td>42</td>\n",
              "      <td>0.93%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15560_e0</td>\n",
              "      <td>EX_cpd15560_e0</td>\n",
              "      <td>0.4434</td>\n",
              "      <td>49</td>\n",
              "      <td>0.10%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd17041_c0</td>\n",
              "      <td>rxn13782_c0</td>\n",
              "      <td>61.65</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd17042_c0</td>\n",
              "      <td>rxn13783_c0</td>\n",
              "      <td>61.65</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd17043_c0</td>\n",
              "      <td>rxn13784_c0</td>\n",
              "      <td>61.65</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>cpd00001_e0</td>\n",
              "      <td>EX_cpd00001_e0</td>\n",
              "      <td>-1000</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00007_e0</td>\n",
              "      <td>EX_cpd00007_e0</td>\n",
              "      <td>-0.1108</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00011_e0</td>\n",
              "      <td>EX_cpd00011_e0</td>\n",
              "      <td>-789.3</td>\n",
              "      <td>1</td>\n",
              "      <td>3.76%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00013_e0</td>\n",
              "      <td>EX_cpd00013_e0</td>\n",
              "      <td>-217.9</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00027_e0</td>\n",
              "      <td>EX_cpd00027_e0</td>\n",
              "      <td>-1000</td>\n",
              "      <td>6</td>\n",
              "      <td>28.60%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00100_e0</td>\n",
              "      <td>EX_cpd00100_e0</td>\n",
              "      <td>-1000</td>\n",
              "      <td>3</td>\n",
              "      <td>14.30%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00122_e0</td>\n",
              "      <td>EX_cpd00122_e0</td>\n",
              "      <td>-975</td>\n",
              "      <td>8</td>\n",
              "      <td>37.18%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00129_e0</td>\n",
              "      <td>EX_cpd00129_e0</td>\n",
              "      <td>-318.3</td>\n",
              "      <td>5</td>\n",
              "      <td>7.59%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00175_e0</td>\n",
              "      <td>EX_cpd00175_e0</td>\n",
              "      <td>-104.4</td>\n",
              "      <td>17</td>\n",
              "      <td>8.46%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd03422_e0</td>\n",
              "      <td>EX_cpd03422_e0</td>\n",
              "      <td>-0.4434</td>\n",
              "      <td>48</td>\n",
              "      <td>0.10%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd11416_c0</td>\n",
              "      <td>SK_cpd11416_c0</td>\n",
              "      <td>-61.65</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ],
      "source": [
        "gf_model.summary()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "VlYo8HCDcSAV",
        "outputId": "26108e80-8e18-46a1-fdd0-40fe3ed2db39",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of reactions added: 81\n"
          ]
        }
      ],
      "source": [
        "print(\"Number of reactions added:\", len(gapfill_complete_medium.added_reactions))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "X0hr1bhCcSAW"
      },
      "source": [
        "In principle this means we are done, let get a beer :)\n",
        "\n",
        "But maybe you still want to learn something about how DNNGIOR works and what else we can do."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3bCJ5S4gcSAW"
      },
      "source": [
        "### How does DNNGIOR work\n",
        "\n",
        "DNNGIOR gapfilling takes two steps:\n",
        "1. Make a prediction based on the reactions in the draft model\n",
        "2. Use these predictions to find weigh the solutions of the linear programming algorythm\n",
        "\n",
        "Lets have a look at these predictions, we can load in the neural network:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "wL1_Cnd5cSAX"
      },
      "outputs": [],
      "source": [
        "NN_MSEED = NN(path=TRAINED_NN_MSEED)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0MLcVfEXcSAX"
      },
      "source": [
        "and make a prediction for our model"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "DYpjf-oXcSAY",
        "outputId": "f46a3796-ccb1-41c8-f309-6da7650f4b07",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#reactions not found in NN-keys:  65 / 726\n"
          ]
        }
      ],
      "source": [
        "prediction = NN_MSEED.predict(draft_model)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_LXoBsSYcSAY"
      },
      "source": [
        "This will give us a prediction for what reactions should be present in the gapfilled model."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "G6dJMAgtcSAY",
        "outputId": "7b48459e-efc2-4c2c-dbf5-760016684f37",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 447
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Axes: ylabel='Frequency'>"
            ]
          },
          "metadata": {},
          "execution_count": 12
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkQAAAGdCAYAAADzOWwgAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAMc5JREFUeJzt3X1clHW+//E3iANqMogu4JzQzCxv002LKG0zOeJNdqN7ypWMWja7gW6kLD3lzWZFUVlqlFvHxE60luekp8xQwopK8gYlTQ0r3bB0oD0oI7Rye/3+6Md1dhILpmFm8Ho9H4/r8dj5fj/XXJ/ru+S8H9dcMxNkGIYhAAAACwv2dwMAAAD+RiACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWF+LvBtqLxsZGHT58WF27dlVQUJC/2wEAAC1gGIaOHz8uh8Oh4OBTXwciELXQ4cOHFRsb6+82AACABw4dOqQzzzzzlPMEohbq2rWrpB8XNDw83M/dAACAlnC5XIqNjTVfx0+FQNRCTW+ThYeHE4gAAGhnful2F26qBgAAlkcgAgAAlkcgAgAAlufXQFRQUKBJkybJ4XAoKChIa9euPalm3759uuqqq2S329WlSxddeOGFKi0tNedPnDih1NRUde/eXWeccYamTJmisrIyt+coLS3VxIkT1blzZ0VFRWnWrFmqr69v69MDAADthF8DUXV1tYYOHaqsrKxm57/++muNHDlS/fv31wcffKBdu3Zp7ty5CgsLM2tmzpypt99+W6tXr9aHH36ow4cPa/LkyeZ8Q0ODJk6cqNraWm3evFkrV65Udna25s2b1+bnBwAA2ocgwzAMfzch/Xj395o1a3TNNdeYY1OnTlXHjh31n//5n83uU1lZqd/85jd67bXX9Pvf/16S9MUXX2jAgAEqLCzUxRdfrHfffVdXXnmlDh8+rOjoaEnSsmXL9MADD+j777+XzWZrUX8ul0t2u12VlZV8ygwAgHaipa/fAXsPUWNjo9555x2de+65SkxMVFRUlOLi4tzeVisqKlJdXZ0SEhLMsf79+6tXr14qLCyUJBUWFmrIkCFmGJKkxMREuVwu7dmz55THr6mpkcvlctsAAMDpKWADUXl5uaqqqvT4449r3Lhx2rhxo6699lpNnjxZH374oSTJ6XTKZrMpIiLCbd/o6Gg5nU6z5p/DUNN809ypZGRkyG63mxvfUg0AwOkrYANRY2OjJOnqq6/WzJkzNWzYMM2ePVtXXnmlli1b1ubHnzNnjiorK83t0KFDbX5MAADgHwEbiHr06KGQkBANHDjQbXzAgAHmp8xiYmJUW1urY8eOudWUlZUpJibGrPnpp86aHjfVNCc0NNT8Vmq+nRoAgNNbwAYim82mCy+8UCUlJW7j+/fvV+/evSVJw4cPV8eOHZWfn2/Ol5SUqLS0VPHx8ZKk+Ph47d69W+Xl5WZNXl6ewsPDTwpbAADAmvz6W2ZVVVX66quvzMcHDx5UcXGxIiMj1atXL82aNUvXX3+9LrvsMo0ePVq5ubl6++239cEHH0iS7Ha7UlJSlJ6ersjISIWHh+vOO+9UfHy8Lr74YknS2LFjNXDgQE2fPl2ZmZlyOp166KGHlJqaqtDQUH+cNgAACDSGH73//vuGpJO25ORks2b58uXGOeecY4SFhRlDhw411q5d6/Yc//jHP4w77rjD6Natm9G5c2fj2muvNY4cOeJW87e//c0YP3680alTJ6NHjx7Gvffea9TV1bWq18rKSkOSUVlZ6fH5AgAA32rp63fAfA9RoON7iAAAaH9a+vrt17fM8P8tWNCyMQAA0CYC9qZqAAAAXyEQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAyyMQAQAAy/NrICooKNCkSZPkcDgUFBSktWvXnrL2tttuU1BQkJ599lm38YqKCiUlJSk8PFwRERFKSUlRVVWVW82uXbs0atQohYWFKTY2VpmZmW1wNgAAoL3yayCqrq7W0KFDlZWV9bN1a9as0aeffiqHw3HSXFJSkvbs2aO8vDytW7dOBQUFmjFjhjnvcrk0duxY9e7dW0VFRXryySe1YMECvfjii14/HwAA0D6F+PPg48eP1/jx43+25rvvvtOdd96pDRs2aOLEiW5z+/btU25urrZt26YRI0ZIkpYuXaoJEyboqaeeksPhUE5Ojmpra/Xyyy/LZrNp0KBBKi4u1qJFi9yCEwAAsK6AvoeosbFR06dP16xZszRo0KCT5gsLCxUREWGGIUlKSEhQcHCwtmzZYtZcdtllstlsZk1iYqJKSkp09OjRUx67pqZGLpfLbQMAAKengA5ETzzxhEJCQnTXXXc1O+90OhUVFeU2FhISosjISDmdTrMmOjrarabpcVNNczIyMmS3280tNjb215wKAAAIYAEbiIqKirR48WJlZ2crKCjI58efM2eOKisrze3QoUM+7wEAAPhGwAaijz76SOXl5erVq5dCQkIUEhKib775Rvfee6/OOussSVJMTIzKy8vd9quvr1dFRYViYmLMmrKyMreapsdNNc0JDQ1VeHi42wYAAE5PARuIpk+frl27dqm4uNjcHA6HZs2apQ0bNkiS4uPjdezYMRUVFZn7bdq0SY2NjYqLizNrCgoKVFdXZ9bk5eXpvPPOU7du3Xx7UgAAICD59VNmVVVV+uqrr8zHBw8eVHFxsSIjI9WrVy91797drb5jx46KiYnReeedJ0kaMGCAxo0bp1tuuUXLli1TXV2d0tLSNHXqVPMj+tOmTdOf//xnpaSk6IEHHtDnn3+uxYsX65lnnvHdiQIAgIDm10C0fft2jR492nycnp4uSUpOTlZ2dnaLniMnJ0dpaWkaM2aMgoODNWXKFC1ZssSct9vt2rhxo1JTUzV8+HD16NFD8+bN4yP3AADAFGQYhuHvJtoDl8slu92uyspK799PtGBBy8YAAECrtPT1O2DvIQIAAPAVAhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8AhEAALA8vwaigoICTZo0SQ6HQ0FBQVq7dq05V1dXpwceeEBDhgxRly5d5HA4dOONN+rw4cNuz1FRUaGkpCSFh4crIiJCKSkpqqqqcqvZtWuXRo0apbCwMMXGxiozM9MXpwcAANoJvwai6upqDR06VFlZWSfN/fDDD9qxY4fmzp2rHTt26M0331RJSYmuuuoqt7qkpCTt2bNHeXl5WrdunQoKCjRjxgxz3uVyaezYserdu7eKior05JNPasGCBXrxxRfb/PwAAED7EOLPg48fP17jx49vds5utysvL89t7LnnntNFF12k0tJS9erVS/v27VNubq62bdumESNGSJKWLl2qCRMm6KmnnpLD4VBOTo5qa2v18ssvy2azadCgQSouLtaiRYvcghMAALCudnUPUWVlpYKCghQRESFJKiwsVEREhBmGJCkhIUHBwcHasmWLWXPZZZfJZrOZNYmJiSopKdHRo0dPeayamhq5XC63DQAAnJ7aTSA6ceKEHnjgAf3hD39QeHi4JMnpdCoqKsqtLiQkRJGRkXI6nWZNdHS0W03T46aa5mRkZMhut5tbbGysN08HAAAEkHYRiOrq6nTdddfJMAy98MILPjnmnDlzVFlZaW6HDh3yyXEBAIDv+fUeopZoCkPffPONNm3aZF4dkqSYmBiVl5e71dfX16uiokIxMTFmTVlZmVtN0+OmmuaEhoYqNDTUW6cBAAACWEBfIWoKQ19++aXee+89de/e3W0+Pj5ex44dU1FRkTm2adMmNTY2Ki4uzqwpKChQXV2dWZOXl6fzzjtP3bp1882JAACAgObXQFRVVaXi4mIVFxdLkg4ePKji4mKVlpaqrq5Ov//977V9+3bl5OSooaFBTqdTTqdTtbW1kqQBAwZo3LhxuuWWW7R161Z98sknSktL09SpU+VwOCRJ06ZNk81mU0pKivbs2aPXX39dixcvVnp6ur9OGwAABBi/vmW2fft2jR492nzcFFKSk5O1YMECvfXWW5KkYcOGue33/vvv6/LLL5ck5eTkKC0tTWPGjFFwcLCmTJmiJUuWmLV2u10bN25Uamqqhg8frh49emjevHl85B4AAJj8Goguv/xyGYZxyvmfm2sSGRmp11577Wdrzj//fH300Uet7g8AAFhDQN9DBAAA4AsEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHkEIgAAYHl+DUQFBQWaNGmSHA6HgoKCtHbtWrd5wzA0b9489ezZU506dVJCQoK+/PJLt5qKigolJSUpPDxcERERSklJUVVVlVvNrl27NGrUKIWFhSk2NlaZmZltfWoAAKAd8Wsgqq6u1tChQ5WVldXsfGZmppYsWaJly5Zpy5Yt6tKlixITE3XixAmzJikpSXv27FFeXp7WrVungoICzZgxw5x3uVwaO3asevfuraKiIj355JNasGCBXnzxxTY/PwAA0D6E+PPg48eP1/jx45udMwxDzz77rB566CFdffXVkqRXXnlF0dHRWrt2raZOnap9+/YpNzdX27Zt04gRIyRJS5cu1YQJE/TUU0/J4XAoJydHtbW1evnll2Wz2TRo0CAVFxdr0aJFbsEJAABYV8DeQ3Tw4EE5nU4lJCSYY3a7XXFxcSosLJQkFRYWKiIiwgxDkpSQkKDg4GBt2bLFrLnssstks9nMmsTERJWUlOjo0aM+OhsAABDI/HqF6Oc4nU5JUnR0tNt4dHS0Oed0OhUVFeU2HxISosjISLeaPn36nPQcTXPdunVr9vg1NTWqqakxH7tcrl9xNgAAIJAF7BUif8vIyJDdbje32NhYf7cEAADaSMAGopiYGElSWVmZ23hZWZk5FxMTo/Lycrf5+vp6VVRUuNU09xz/fIzmzJkzR5WVleZ26NChX3dCAAAgYAVsIOrTp49iYmKUn59vjrlcLm3ZskXx8fGSpPj4eB07dkxFRUVmzaZNm9TY2Ki4uDizpqCgQHV1dWZNXl6ezjvvvFO+XSZJoaGhCg8Pd9sAAMDpya+BqKqqSsXFxSouLpb0443UxcXFKi0tVVBQkO655x498sgjeuutt7R7927deOONcjgcuuaaayRJAwYM0Lhx43TLLbdo69at+uSTT5SWlqapU6fK4XBIkqZNmyabzaaUlBTt2bNHr7/+uhYvXqz09HQ/nTUAAAg0fr2pevv27Ro9erT5uCmkJCcnKzs7W/fff7+qq6s1Y8YMHTt2TCNHjlRubq7CwsLMfXJycpSWlqYxY8YoODhYU6ZM0ZIlS8x5u92ujRs3KjU1VcOHD1ePHj00b948PnIPAABMQYZhGP5uoj1wuVyy2+2qrKz0/ttnCxa0bAwAALRKS1+/A/YeIgAAAF/xKBAdOHDA230AAAD4jUeB6JxzztHo0aP16quvuv2uGAAAQHvkUSDasWOHzj//fKWnpysmJka33nqrtm7d6u3eAAAAfMKjQDRs2DAtXrxYhw8f1ssvv6wjR45o5MiRGjx4sBYtWqTvv//e230CAAC0mV91U3VISIgmT56s1atX64knntBXX32l++67T7Gxsbrxxht15MgRb/UJAADQZn5VINq+fbvuuOMO9ezZU4sWLdJ9992nr7/+Wnl5eTp8+LCuvvpqb/UJAADQZjz6YsZFixZpxYoVKikp0YQJE/TKK69owoQJCg7+MV/16dNH2dnZOuuss7zZKwAAQJvwKBC98MIL+uMf/6ibbrpJPXv2bLYmKipKy5cv/1XNAQAA+IJHgejLL7/8xRqbzabk5GRPnh4AAMCnPLqHaMWKFVq9evVJ46tXr9bKlSt/dVMAAAC+5FEgysjIUI8ePU4aj4qK0mOPPfarmwIAAPAljwJRaWmp+vTpc9J47969VVpa+qubAgAA8CWPAlFUVJR27dp10vhnn32m7t27/+qmAAAAfMmjQPSHP/xBd911l95//301NDSooaFBmzZt0t13362pU6d6u0cAAIA25dGnzBYuXKi//e1vGjNmjEJCfnyKxsZG3XjjjdxDBAAA2h2PApHNZtPrr7+uhQsX6rPPPlOnTp00ZMgQ9e7d29v9AQAAtDmPAlGTc889V+eee663egEAAPALjwJRQ0ODsrOzlZ+fr/LycjU2NrrNb9q0ySvNAQAA+IJHgejuu+9Wdna2Jk6cqMGDBysoKMjbfQEAAPiMR4Fo1apVeuONNzRhwgRv9wMAAOBzHn3s3maz6ZxzzvF2LwAAAH7hUSC69957tXjxYhmG4e1+AAAAfM6jt8w+/vhjvf/++3r33Xc1aNAgdezY0W3+zTff9EpzAAAAvuBRIIqIiNC1117r7V4AAAD8wqNAtGLFCm/3AQAA4Dce3UMkSfX19Xrvvff0l7/8RcePH5ckHT58WFVVVV5rDgAAwBc8ukL0zTffaNy4cSotLVVNTY3+9V//VV27dtUTTzyhmpoaLVu2zNt9AgAAtBmPrhDdfffdGjFihI4ePapOnTqZ49dee63y8/O91hwAAIAveHSF6KOPPtLmzZtls9ncxs866yx99913XmkMAADAVzy6QtTY2KiGhoaTxr/99lt17dr1VzcFAADgSx4ForFjx+rZZ581HwcFBamqqkrz58/n5zwAAEC749FbZk8//bQSExM1cOBAnThxQtOmTdOXX36pHj166K9//au3ewQAAGhTHgWiM888U5999plWrVqlXbt2qaqqSikpKUpKSnK7yRoAAKA98Ph7iEJCQnTDDTcoMzNTzz//vP70pz95PQw1NDRo7ty56tOnjzp16qS+fftq4cKFbr+hZhiG5s2bp549e6pTp05KSEjQl19+6fY8FRUVSkpKUnh4uCIiIpSSksL3JQEAAJNHV4heeeWVn52/8cYbPWrmp5544gm98MILWrlypQYNGqTt27fr5ptvlt1u11133SVJyszM1JIlS7Ry5Ur16dNHc+fOVWJiovbu3auwsDBJUlJSko4cOaK8vDzV1dXp5ptv1owZM/Taa695pU8AANC+BRke/GR9t27d3B7X1dXphx9+kM1mU+fOnVVRUeGV5q688kpFR0dr+fLl5tiUKVPUqVMnvfrqqzIMQw6HQ/fee6/uu+8+SVJlZaWio6OVnZ2tqVOnat++fRo4cKC2bdumESNGSJJyc3M1YcIEffvtt3I4HC3qxeVyyW63q7KyUuHh4V45P9OCBS0bAwAArdLS12+P3jI7evSo21ZVVaWSkhKNHDnSqzdVX3LJJcrPz9f+/fslSZ999pk+/vhjjR8/XpJ08OBBOZ1OJSQkmPvY7XbFxcWpsLBQklRYWKiIiAgzDElSQkKCgoODtWXLFq/1CgAA2i+P3jJrTr9+/fT444/rhhtu0BdffOGV55w9e7ZcLpf69++vDh06qKGhQY8++qiSkpIkSU6nU5IUHR3ttl90dLQ553Q6FRUV5TYfEhKiyMhIs6Y5NTU1qqmpMR+7XC6vnBMAAAg8Ht9U3ZyQkBAdPnzYa8/3xhtvKCcnR6+99pp27NihlStX6qmnntLKlSu9doxTycjIkN1uN7fY2Ng2PyYAAPAPj64QvfXWW26PDcPQkSNH9Nxzz+nSSy/1SmOSNGvWLM2ePVtTp06VJA0ZMkTffPONMjIylJycrJiYGElSWVmZevbsae5XVlamYcOGSZJiYmJUXl7u9rz19fWqqKgw92/OnDlzlJ6ebj52uVyEIgAATlMeBaJrrrnG7XFQUJB+85vf6IorrtDTTz/tjb4kST/88IOCg90vYnXo0EGNjY2SpD59+igmJkb5+flmAHK5XNqyZYtuv/12SVJ8fLyOHTumoqIiDR8+XJK0adMmNTY2Ki4u7pTHDg0NVWhoqNfOBQAABC6PAlFTIGlrkyZN0qOPPqpevXpp0KBB2rlzpxYtWqQ//vGPkn4MYvfcc48eeeQR9evXz/zYvcPhMEPbgAEDNG7cON1yyy1atmyZ6urqlJaWpqlTp7b4E2YAAOD05rWbqtvC0qVLNXfuXN1xxx0qLy+Xw+HQrbfeqnnz5pk1999/v6qrqzVjxgwdO3ZMI0eOVG5urvkdRJKUk5OjtLQ0jRkzRsHBwZoyZYqWLFnij1MCAAAByKPvIfrne2t+yaJFi1r79AGJ7yECAKD9aenrt0dXiHbu3KmdO3eqrq5O5513niRp//796tChgy644AKzLigoyJOnBwAA8CmPAtGkSZPUtWtXrVy50vzW6qNHj+rmm2/WqFGjdO+993q1SQAAgLbk0fcQPf3008rIyHD7CY9u3brpkUce8eqnzAAAAHzBo0Dkcrn0/fffnzT+/fff6/jx47+6KQAAAF/yKBBde+21uvnmm/Xmm2/q22+/1bfffqv//u//VkpKiiZPnuztHgEAANqUR/cQLVu2TPfdd5+mTZumurq6H58oJEQpKSl68sknvdogAABAW/MoEHXu3FnPP/+8nnzySX399deSpL59+6pLly5ebQ4AAMAXftWPux45ckRHjhxRv3791KVLF3nwlUYAAAB+51Eg+t///V+NGTNG5557riZMmKAjR45IklJSUvjIPQAAaHc8CkQzZ85Ux44dVVpaqs6dO5vj119/vXJzc73WHAAAgC94dA/Rxo0btWHDBp155plu4/369dM333zjlcYAAAB8xaMrRNXV1W5XhppUVFQoNDT0VzcFAADgSx4FolGjRumVV14xHwcFBamxsVGZmZkaPXq015oDAADwBY/eMsvMzNSYMWO0fft21dbW6v7779eePXtUUVGhTz75xNs9AgAAtCmPrhANHjxY+/fv18iRI3X11VerurpakydP1s6dO9W3b19v9wgAANCmWn2FqK6uTuPGjdOyZcv04IMPtkVPAAAAPtXqK0QdO3bUrl272qIXAAAAv/DoLbMbbrhBy5cv93YvAAAAfuHRTdX19fV6+eWX9d5772n48OEn/YbZokWLvNIcAACAL7QqEB04cEBnnXWWPv/8c11wwQWSpP3797vVBAUFea87AAAAH2hVIOrXr5+OHDmi999/X9KPP9WxZMkSRUdHt0lzAAAAvtCqe4h++mv27777rqqrq73aEAAAgK95dFN1k58GJAAAgPaoVYEoKCjopHuEuGcIAAC0d626h8gwDN10003mD7ieOHFCt91220mfMnvzzTe91yEAAEAba1UgSk5Odnt8ww03eLUZAAAAf2hVIFqxYkVb9QEAAOA3v+qmagAAgNMBgQgAAFgegQgAAFgegQgAAFgegQgAAFgegQgAAFgegQgAAFhewAei7777TjfccIO6d++uTp06aciQIdq+fbs5bxiG5s2bp549e6pTp05KSEjQl19+6fYcFRUVSkpKUnh4uCIiIpSSkqKqqipfnwoAAAhQAR2Ijh49qksvvVQdO3bUu+++q7179+rpp59Wt27dzJrMzEwtWbJEy5Yt05YtW9SlSxclJibqxIkTZk1SUpL27NmjvLw8rVu3TgUFBZoxY4Y/TgkAAASgICOAf7J+9uzZ+uSTT/TRRx81O28YhhwOh+69917dd999kqTKykpFR0crOztbU6dO1b59+zRw4EBt27ZNI0aMkCTl5uZqwoQJ+vbbb+VwOFrUi8vlkt1uV2VlpcLDw71zgk0WLGjZGAAAaJWWvn4H9BWit956SyNGjNC//du/KSoqSr/97W/10ksvmfMHDx6U0+lUQkKCOWa32xUXF6fCwkJJUmFhoSIiIswwJEkJCQkKDg7Wli1bTnnsmpoauVwutw0AAJyeAjoQHThwQC+88IL69eunDRs26Pbbb9ddd92llStXSpKcTqckKTo62m2/6Ohoc87pdCoqKsptPiQkRJGRkWZNczIyMmS3280tNjbWm6cGAAACSEAHosbGRl1wwQV67LHH9Nvf/lYzZszQLbfcomXLlrX5sefMmaPKykpzO3ToUJsfEwAA+EdAB6KePXtq4MCBbmMDBgxQaWmpJCkmJkaSVFZW5lZTVlZmzsXExKi8vNxtvr6+XhUVFWZNc0JDQxUeHu62AQCA01NAB6JLL71UJSUlbmP79+9X7969JUl9+vRRTEyM8vPzzXmXy6UtW7YoPj5ekhQfH69jx46pqKjIrNm0aZMaGxsVFxfng7MAAACBLsTfDfycmTNn6pJLLtFjjz2m6667Tlu3btWLL76oF198UZIUFBSke+65R4888oj69eunPn36aO7cuXI4HLrmmmsk/XhFady4ceZbbXV1dUpLS9PUqVNb/AkzAABwegvoQHThhRdqzZo1mjNnjh5++GH16dNHzz77rJKSksya+++/X9XV1ZoxY4aOHTumkSNHKjc3V2FhYWZNTk6O0tLSNGbMGAUHB2vKlClasmSJP04JAAAEoID+HqJAwvcQAQDQ/pwW30MEAADgCwQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgeQQiAABgee0qED3++OMKCgrSPffcY46dOHFCqamp6t69u8444wxNmTJFZWVlbvuVlpZq4sSJ6ty5s6KiojRr1izV19f7uHsAABCo2k0g2rZtm/7yl7/o/PPPdxufOXOm3n77ba1evVoffvihDh8+rMmTJ5vzDQ0Nmjhxompra7V582atXLlS2dnZmjdvnq9PAQAABKh2EYiqqqqUlJSkl156Sd26dTPHKysrtXz5ci1atEhXXHGFhg8frhUrVmjz5s369NNPJUkbN27U3r179eqrr2rYsGEaP368Fi5cqKysLNXW1vrrlAAAQABpF4EoNTVVEydOVEJCgtt4UVGR6urq3Mb79++vXr16qbCwUJJUWFioIUOGKDo62qxJTEyUy+XSnj17TnnMmpoauVwutw0AAJyeQvzdwC9ZtWqVduzYoW3btp0053Q6ZbPZFBER4TYeHR0tp9Np1vxzGGqab5o7lYyMDP35z3/+ld0DAID2IKCvEB06dEh33323cnJyFBYW5tNjz5kzR5WVleZ26NAhnx4fAAD4TkBfISoqKlJ5ebkuuOACc6yhoUEFBQV67rnntGHDBtXW1urYsWNuV4nKysoUExMjSYqJidHWrVvdnrfpU2hNNc0JDQ1VaGioF88GAABIkhYsaNmYDwX0FaIxY8Zo9+7dKi4uNrcRI0YoKSnJ/N8dO3ZUfn6+uU9JSYlKS0sVHx8vSYqPj9fu3btVXl5u1uTl5Sk8PFwDBw70+TkBAIDAE9BXiLp27arBgwe7jXXp0kXdu3c3x1NSUpSenq7IyEiFh4frzjvvVHx8vC6++GJJ0tixYzVw4EBNnz5dmZmZcjqdeuihh5SamsoVIAAAICnAA1FLPPPMMwoODtaUKVNUU1OjxMREPf/88+Z8hw4dtG7dOt1+++2Kj49Xly5dlJycrIcfftiPXQMAgEDS7gLRBx984PY4LCxMWVlZysrKOuU+vXv31vr169u4MwAA0F4F9D1EAAAAvkAgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlkcgAgAAlhfwgSgjI0MXXnihunbtqqioKF1zzTUqKSlxqzlx4oRSU1PVvXt3nXHGGZoyZYrKysrcakpLSzVx4kR17txZUVFRmjVrlurr6315KgAAIEAFfCD68MMPlZqaqk8//VR5eXmqq6vT2LFjVV1dbdbMnDlTb7/9tlavXq0PP/xQhw8f1uTJk835hoYGTZw4UbW1tdq8ebNWrlyp7OxszZs3zx+nBAAAAkyIvxv4Jbm5uW6Ps7OzFRUVpaKiIl122WWqrKzU8uXL9dprr+mKK66QJK1YsUIDBgzQp59+qosvvlgbN27U3r179d577yk6OlrDhg3TwoUL9cADD2jBggWy2Wz+ODUAABAgAv4K0U9VVlZKkiIjIyVJRUVFqqurU0JCglnTv39/9erVS4WFhZKkwsJCDRkyRNHR0WZNYmKiXC6X9uzZ0+xxampq5HK53DYAAHB6aleBqLGxUffcc48uvfRSDR48WJLkdDpls9kUERHhVhsdHS2n02nW/HMYappvmmtORkaG7Ha7ucXGxnr5bAAAQKBoV4EoNTVVn3/+uVatWtXmx5ozZ44qKyvN7dChQ21+TAAA4B8Bfw9Rk7S0NK1bt04FBQU688wzzfGYmBjV1tbq2LFjbleJysrKFBMTY9Zs3brV7fmaPoXWVPNToaGhCg0N9fJZAACAQBTwV4gMw1BaWprWrFmjTZs2qU+fPm7zw4cPV8eOHZWfn2+OlZSUqLS0VPHx8ZKk+Ph47d69W+Xl5WZNXl6ewsPDNXDgQN+cCAAACFgBf4UoNTVVr732mv7nf/5HXbt2Ne/5sdvt6tSpk+x2u1JSUpSenq7IyEiFh4frzjvvVHx8vC6++GJJ0tixYzVw4EBNnz5dmZmZcjqdeuihh5SamspVIAAAEPiB6IUXXpAkXX755W7jK1as0E033SRJeuaZZxQcHKwpU6aopqZGiYmJev75583aDh06aN26dbr99tsVHx+vLl26KDk5WQ8//LCvTgMAAASwgA9EhmH8Yk1YWJiysrKUlZV1yprevXtr/fr13mwNAACcJgL+HiIAAIC2RiACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWF+LvBnAKCxb8/GMAAOA1XCECAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWRyACAACWx2+ZAQCAttUOfo+TQNReNPfH1A7+wAAAp7nT5MfICUTt2WnyRwgAgL9ZKhBlZWXpySeflNPp1NChQ7V06VJddNFF/m7Le7iKBADwt3b6umOZm6pff/11paena/78+dqxY4eGDh2qxMRElZeX+7s1AADgZ0GGYRj+bsIX4uLidOGFF+q5556TJDU2Nio2NlZ33nmnZs+e/Yv7u1wu2e12VVZWKjw83LvN+TJNt+RYLbnS5OnVKG/VAABazlvvILTlv89t9Nwtff22xFtmtbW1Kioq0pw5c8yx4OBgJSQkqLCwsNl9ampqVFNTYz6urKyU9OPCet0/HafNNdd/Rkbr92uu55/WtOR5W3IsT/30+P/0/79fBFo/sKbm/rsM9L9FT/8t8Zafro8v19DTc/9pP839mx1o/7+3xeur/u91+xev/xgW8N133xmSjM2bN7uNz5o1y7joooua3Wf+/PmGJDY2NjY2NrbTYDt06NDPZgVLXCHyxJw5c5Senm4+bmxsVEVFhbp3766goCCvHcflcik2NlaHDh3y/ltxMLHOvsE6+w5r7Russ2+05TobhqHjx4/L4XD8bJ0lAlGPHj3UoUMHlZWVuY2XlZUpJiam2X1CQ0MVGhrqNhYREdFWLSo8PJz/2HyAdfYN1tl3WGvfYJ19o63W2W63/2KNJT5lZrPZNHz4cOXn55tjjY2Nys/PV3x8vB87AwAAgcASV4gkKT09XcnJyRoxYoQuuugiPfvss6qurtbNN9/s79YAAICfWSYQXX/99fr+++81b948OZ1ODRs2TLm5uYqOjvZrX6GhoZo/f/5Jb8/Bu1hn32CdfYe19g3W2TcCYZ0t8z1EAAAAp2KJe4gAAAB+DoEIAABYHoEIAABYHoEIAABYHoHIB7KysnTWWWcpLCxMcXFx2rp168/Wr169Wv3791dYWJiGDBmi9evX+6jT9q016/zSSy9p1KhR6tatm7p166aEhIRf/P8FP2rt33OTVatWKSgoSNdcc03bNniaaO06Hzt2TKmpqerZs6dCQ0N17rnn8m9HC7V2rZ999lmdd9556tSpk2JjYzVz5kydOHHCR922TwUFBZo0aZIcDoeCgoK0du3aX9zngw8+0AUXXKDQ0FCdc845ys7ObtsmvfNrYTiVVatWGTabzXj55ZeNPXv2GLfccosRERFhlJWVNVv/ySefGB06dDAyMzONvXv3Gg899JDRsWNHY/fu3T7uvH1p7TpPmzbNyMrKMnbu3Gns27fPuOmmmwy73W58++23Pu68fWntOjc5ePCg8S//8i/GqFGjjKuvvto3zbZjrV3nmpoaY8SIEcaECROMjz/+2Dh48KDxwQcfGMXFxT7uvP1p7Vrn5OQYoaGhRk5OjnHw4EFjw4YNRs+ePY2ZM2f6uPP2Zf369caDDz5ovPnmm4YkY82aNT9bf+DAAaNz585Genq6sXfvXmPp0qVGhw4djNzc3DbrkUDUxi666CIjNTXVfNzQ0GA4HA4jIyOj2frrrrvOmDhxottYXFycceutt7Zpn+1da9f5p+rr642uXbsaK1eubKsWTwuerHN9fb1xySWXGP/xH/9hJCcnE4haoLXr/MILLxhnn322UVtb66sWTxutXevU1FTjiiuucBtLT083Lr300jbt83TSkkB0//33G4MGDXIbu/76643ExMQ264u3zNpQbW2tioqKlJCQYI4FBwcrISFBhYWFze5TWFjoVi9JiYmJp6yHZ+v8Uz/88IPq6uoUGRnZVm22e56u88MPP6yoqCilpKT4os12z5N1fuuttxQfH6/U1FRFR0dr8ODBeuyxx9TQ0OCrttslT9b6kksuUVFRkfm22oEDB7R+/XpNmDDBJz1bhT9eCy3zTdX+8Pe//10NDQ0nfRt2dHS0vvjii2b3cTqdzdY7nc4267O982Sdf+qBBx6Qw+E46T9A/B9P1vnjjz/W8uXLVVxc7IMOTw+erPOBAwe0adMmJSUlaf369frqq690xx13qK6uTvPnz/dF2+2SJ2s9bdo0/f3vf9fIkSNlGIbq6+t122236d///d990bJlnOq10OVy6R//+Ic6derk9WNyhQiW9/jjj2vVqlVas2aNwsLC/N3OaeP48eOaPn26XnrpJfXo0cPf7ZzWGhsbFRUVpRdffFHDhw/X9ddfrwcffFDLli3zd2unnQ8++ECPPfaYnn/+ee3YsUNvvvmm3nnnHS1cuNDfreFX4gpRG+rRo4c6dOigsrIyt/GysjLFxMQ0u09MTEyr6uHZOjd56qmn9Pjjj+u9997T+eef35ZttnutXeevv/5af/vb3zRp0iRzrLGxUZIUEhKikpIS9e3bt22bboc8+Xvu2bOnOnbsqA4dOphjAwYMkNPpVG1trWw2W5v23F55stZz587V9OnT9ac//UmSNGTIEFVXV2vGjBl68MEHFRzMdQZvONVrYXh4eJtcHZK4QtSmbDabhg8frvz8fHOssbFR+fn5io+Pb3af+Ph4t3pJysvLO2U9PFtnScrMzNTChQuVm5urESNG+KLVdq2169y/f3/t3r1bxcXF5nbVVVdp9OjRKi4uVmxsrC/bbzc8+Xu+9NJL9dVXX5mBU5L279+vnj17EoZ+hidr/cMPP5wUepqCqMFPg3qNX14L2+x2bRiG8eNHOkNDQ43s7Gxj7969xowZM4yIiAjD6XQahmEY06dPN2bPnm3Wf/LJJ0ZISIjx1FNPGfv27TPmz5/Px+5boLXr/Pjjjxs2m834r//6L+PIkSPmdvz4cX+dQrvQ2nX+KT5l1jKtXefS0lKja9euRlpamlFSUmKsW7fOiIqKMh555BF/nUK70dq1nj9/vtG1a1fjr3/9q3HgwAFj48aNRt++fY3rrrvOX6fQLhw/ftzYuXOnsXPnTkOSsWjRImPnzp3GN998YxiGYcyePduYPn26Wd/0sftZs2YZ+/btM7KysvjY/elg6dKlRq9evQybzWZcdNFFxqeffmrO/e53vzOSk5Pd6t944w3j3HPPNWw2mzFo0CDjnXfe8XHH7VNr1rl3796GpJO2+fPn+77xdqa1f8//jEDUcq1d582bNxtxcXFGaGiocfbZZxuPPvqoUV9f7+Ou26fWrHVdXZ2xYMECo2/fvkZYWJgRGxtr3HHHHcbRo0d933g78v777zf7b27T2iYnJxu/+93vTtpn2LBhhs1mM84++2xjxYoVbdpjkGFwjQ8AAFgb9xABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADLIxABAADL+3+0uHpfXcK9xQAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        }
      ],
      "source": [
        "pd.Series(prediction).plot.hist(bins=100, alpha=0.5, color='r', label='MSEED')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l5b_e3PFcSAZ"
      },
      "source": [
        "These are part of the microbial reactome (all the reactions that were present), as the network cannot make predictions for reactions it has never seen it will not give a prediction for those. All other reactions will get the default cost which is normally set to 1.0 but can be changed if you want to prioritize reactions from the reactome:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "aMQFLTSmcSAZ",
        "outputId": "5da8874b-8b0b-44bf-c0e1-63e476549fdd",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#reactions not found in NN-keys:  65 / 726\n"
          ]
        }
      ],
      "source": [
        "gapfill_higher_def_cost = Gapfill(draftModel = path_to_draft_model, objectiveName = 'bio1', default_cost=10, gapfill=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {
        "id": "2pPiI1zIcSAa",
        "outputId": "224d2c36-5818-41e3-c243-899d79dac78c",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 447
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Axes: ylabel='Frequency'>"
            ]
          },
          "metadata": {},
          "execution_count": 14
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjoAAAGdCAYAAAAbudkLAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAI2xJREFUeJzt3XtwVOXhxvFnk5AANkQgJSFCCK2gDZdEE0IR+BUkNQYGC0wtVpAYGaaXTQVXtKGOMB0vASwZvByNWAFpqyBTRSuFESMjatEEMFiK3BQkEnJBNEvikODu/v5w2JkYbtlscpZ3v5+Z/eO85/Duk53Wfebse85x+Hw+nwAAAAwUYXcAAACAjkLRAQAAxqLoAAAAY1F0AACAsSg6AADAWBQdAABgLIoOAAAwFkUHAAAYK8ruAHbzer2qqqpSbGysHA6H3XEAAMAl8Pl8OnXqlJKSkhQRcf7zNmFfdKqqqtS/f3+7YwAAgABUVlaqX79+590f9kUnNjZW0ncfVI8ePWxOAwAALoXb7Vb//v393+PnE7ZFx7IsWZYlj8cjSerRowdFBwCAy8zFlp04wv2hnm63W3Fxcaqvr6foAABwmbjU72+uugIAAMai6AAAAGOFbdGxLEupqakaMWKE3VEAAEAHYY0Oa3QAALjssEYHAACEPYoOAAAwFkUHAAAYi6IDAACMFbZFh6uuAAAwH1ddcdUVAACXHa66AgAAYY+iAwAAjBW2Ty8HAADBl1K4scX2kcWTbEryHc7oAAAAY1F0AACAscK26HB5OQAA5gvbouN0OrV3716Vl5fbHQUAAHSQsC06AADAfBQdAABgLIoOAAAwFkUHAAAYi6IDAACMRdEBAADGougAAABjhW3R4YaBAACYL2yLDjcMBADAfGFbdAAAgPkoOgAAwFgUHQAAYCyKDgAAMBZFBwAAGIuiAwAAjEXRAQAAxqLoAAAAY1F0AACAsSg6AADAWBQdAABgrLAtOjzUEwAA84Vt0eGhngAAmC9siw4AADAfRQcAABiLogMAAIxF0QEAAMai6AAAAGNRdAAAgLEoOgAAwFgUHQAAYCyKDgAAMBZFBwAAGIuiAwAAjEXRAQAAxqLoAAAAY1F0AACAsYwpOt98840GDBig+fPn2x0FAACECGOKziOPPKKf/vSndscAAAAhxIiic/DgQe3bt0+5ubl2RwEAACHE9qKzbds2TZ48WUlJSXI4HNqwYUOrYyzLUkpKirp27aqRI0eqrKysxf758+erqKiokxIDAIDLhe1Fp7GxUWlpabIs65z7161bJ5fLpUWLFmnXrl1KS0tTTk6OamtrJUmvvfaaBg8erMGDB3dmbAAAcBmIsjtAbm7uBX9yKi4u1pw5c5Sfny9JKikp0caNG7Vy5UoVFhbqgw8+0Nq1a7V+/Xo1NDTozJkz6tGjhxYuXHjO+ZqamtTU1OTfdrvdwf2DAABAyLD9jM6FNDc3a+fOncrOzvaPRUREKDs7W9u3b5ckFRUVqbKyUkeOHNFf/vIXzZkz57wl5+zxcXFx/lf//v07/O8AAAD2COmic+LECXk8HiUkJLQYT0hIUHV1dUBzLliwQPX19f5XZWVlMKICAIAQZPtPV8F05513XvSYmJgYxcTEdHwYAABgu5A+oxMfH6/IyEjV1NS0GK+pqVFiYmK75rYsS6mpqRoxYkS75gEAAKErpItOdHS0MjIyVFpa6h/zer0qLS3VqFGj2jW30+nU3r17VV5e3t6YAAAgRNn+01VDQ4MOHTrk3z58+LAqKirUq1cvJScny+VyKS8vT5mZmcrKytLy5cvV2NjovwoLAADgfGwvOjt27ND48eP92y6XS5KUl5en1atXa/r06aqrq9PChQtVXV2t9PR0bd68udUCZQAAgO9z+Hw+n90h7GBZlizLksfj0YEDB1RfX68ePXrYHQsAgMtaSuHGFttHFk/qkPdxu92Ki4u76Pd3SK/R6Uis0QEAwHxhW3QAAID5KDoAAMBYYVt0uI8OAADmC9uiwxodAADMF7ZFBwAAmI+iAwAAjEXRAQAAxgrbosNiZAAAzBe2RYfFyAAAmC9siw4AADAfRQcAABiLogMAAIxF0QEAAMYK26LDVVcAAJgvbIsOV10BAGC+sC06AADAfBQdAABgLIoOAAAwFkUHAAAYi6IDAACMFbZFh8vLAQAwX9gWHS4vBwDAfGFbdAAAgPkoOgAAwFgUHQAAYCyKDgAAMBZFBwAAGIuiAwAAjEXRAQAAxgrbosMNAwEAMF/YFh1uGAgAgPnCtugAAADzUXQAAICxKDoAAMBYFB0AAGAsig4AADAWRQcAABiLogMAAIxF0QEAAMai6AAAAGNRdAAAgLHCtujwrCsAAMwXtkWHZ10BAGC+sC06AADAfBQdAABgLIoOAAAwFkUHAAAYK8ruAOg8KYUbW40dWTzJhiQAAHQOzugAAABjUXQAAICxKDoAAMBYFB0AAGAsig4AADAWRQcAABiLogMAAIxF0QEAAMai6AAAAGNd9kXn66+/VmZmptLT0zV06FA999xzdkcCAAAh4rJ/BERsbKy2bdum7t27q7GxUUOHDtW0adPUu3dvu6MBAACbXfZndCIjI9W9e3dJUlNTk3w+n3w+n82pAABAKLC96Gzbtk2TJ09WUlKSHA6HNmzY0OoYy7KUkpKirl27auTIkSorK2ux/+uvv1ZaWpr69eun++67T/Hx8Z2UHgAAhDLbi05jY6PS0tJkWdY5969bt04ul0uLFi3Srl27lJaWppycHNXW1vqPufLKK7V7924dPnxYL774ompqajorPgAACGG2F53c3Fw9/PDDmjp16jn3FxcXa86cOcrPz1dqaqpKSkrUvXt3rVy5stWxCQkJSktL07vvvnve92tqapLb7W7xAgAAZrK96FxIc3Ozdu7cqezsbP9YRESEsrOztX37dklSTU2NTp06JUmqr6/Xtm3bdM0115x3zqKiIsXFxflf/fv379g/AgAA2Caki86JEyfk8XiUkJDQYjwhIUHV1dWSpM8//1xjx45VWlqaxo4dqz/84Q8aNmzYeedcsGCB6uvr/a/KysoO/RsAAIB9LvvLy7OyslRRUXHJx8fExCgmJqbjAgEAgJAR0md04uPjFRkZ2WpxcU1NjRITE21KBQAALhchXXSio6OVkZGh0tJS/5jX61VpaalGjRrVrrkty1JqaqpGjBjR3pgAACBE2f7TVUNDgw4dOuTfPnz4sCoqKtSrVy8lJyfL5XIpLy9PmZmZysrK0vLly9XY2Kj8/Px2va/T6ZTT6ZTb7VZcXFx7/wwAABCCbC86O3bs0Pjx4/3bLpdLkpSXl6fVq1dr+vTpqqur08KFC1VdXa309HRt3ry51QJlAACA77O96IwbN+6ij2woKChQQUFBUN/XsixZliWPxxPUeQEAQOgI6TU6HcnpdGrv3r0qLy+3OwoAAOggYVt0AACA+Sg6AADAWGFbdLi8HAAA84Vt0WGNDgAA5gvbogMAAMxH0QEAAMai6AAAAGOFbdFhMTIAAOYLqOh89tlnwc7R6ViMDACA+QIqOldffbXGjx+vv//97zp9+nSwMwEAAARFQEVn165dGj58uFwulxITE/Wb3/xGZWVlwc4GAADQLgEVnfT0dD3++OOqqqrSypUrdfz4cY0ZM0ZDhw5VcXGx6urqgp0TAACgzdq1GDkqKkrTpk3T+vXrtWTJEh06dEjz589X//79NWvWLB0/fjxYOQEAANqsXUVnx44d+v3vf6++ffuquLhY8+fP16effqotW7aoqqpKv/jFL4KVM+i46goAAPNFBfKPiouLtWrVKu3fv18TJ07UmjVrNHHiREVEfNebBg4cqNWrVyslJSWYWYPK6XTK6XTK7XYrLi7O7jgAAKADBFR0nnnmGd11112688471bdv33Me06dPHz3//PPtCgcAANAeARWdgwcPXvSY6Oho5eXlBTI9AABAUAS0RmfVqlVav359q/H169frhRdeaHcoAACAYAio6BQVFSk+Pr7VeJ8+ffToo4+2OxQAAEAwBFR0jh49qoEDB7YaHzBggI4ePdruUAAAAMEQUNHp06ePPv7441bju3fvVu/evdsdqjNweTkAAOYLqOj8+te/1t13362tW7fK4/HI4/Ho7bff1ty5c3XbbbcFO2OH4KGeAACYL6Crrh566CEdOXJEEyZMUFTUd1N4vV7NmjWLNToAACBkBFR0oqOjtW7dOj300EPavXu3unXrpmHDhmnAgAHBzgcAABCwgIrOWYMHD9bgwYODlQUAACCoAio6Ho9Hq1evVmlpqWpra+X1elvsf/vtt4MSDgAAoD0CKjpz587V6tWrNWnSJA0dOlQOhyPYuQAAANotoKKzdu1avfzyy5o4cWKw8wAAAARNQJeXR0dH6+qrrw52FgAAgKAKqOjce++9evzxx+Xz+YKdp9Nww0AAAMwX0E9X7733nrZu3apNmzZpyJAh6tKlS4v9r7zySlDCdSSn0ymn0ym32624uDi74wAAgA4QUNG58sorNXXq1GBnAQAACKqAis6qVauCnQMAACDoAlqjI0nffvut3nrrLT377LM6deqUJKmqqkoNDQ1BCwcAANAeAZ3R+fzzz3XzzTfr6NGjampq0s9//nPFxsZqyZIlampqUklJSbBzAgAAtFlAZ3Tmzp2rzMxMffXVV+rWrZt/fOrUqSotLQ1aOAAAgPYI6IzOu+++q//85z+Kjo5uMZ6SkqJjx44FJRgAAEB7BXRGx+v1yuPxtBr/4osvFBsb2+5QAAAAwRBQ0bnpppu0fPly/7bD4VBDQ4MWLVrEYyEAAEDICOinq2XLliknJ0epqak6ffq0br/9dh08eFDx8fF66aWXgp0RAAAgIAEVnX79+mn37t1au3atPv74YzU0NGj27NmaMWNGi8XJAAAAdgqo6EhSVFSUZs6cGcwsAAAAQRVQ0VmzZs0F98+aNSugMJ3JsixZlnXORdUAAMAMARWduXPnttg+c+aMvvnmG0VHR6t79+6XRdHhoZ4AAJgvoKuuvvrqqxavhoYG7d+/X2PGjGExMgAACBkBP+vq+wYNGqTFixe3OtsDAABgl6AVHem7BcpVVVXBnBIAACBgAa3Ref3111ts+3w+HT9+XE899ZRGjx4dlGAAAADtFVDRmTJlSotth8OhH/7wh7rxxhu1bNmyYOQCAABot4CKjtfrDXYOAACAoAvqGh0AAIBQEtAZHZfLdcnHFhcXB/IWAAAA7RZQ0fnoo4/00Ucf6cyZM7rmmmskSQcOHFBkZKSuv/56/3EOhyM4KQEAAAIQUNGZPHmyYmNj9cILL6hnz56SvruJYH5+vsaOHat77703qCEBAAACEdAanWXLlqmoqMhfciSpZ8+eevjhh7nqCgAAhIyAio7b7VZdXV2r8bq6Op06dardoQAAAIIhoKIzdepU5efn65VXXtEXX3yhL774Qv/85z81e/ZsTZs2LdgZAQAAAhLQGp2SkhLNnz9ft99+u86cOfPdRFFRmj17th577LGgBgQAAAhUQGd0unfvrqefflpffvml/wqskydP6umnn9YVV1wR7IwXVFlZqXHjxik1NVXDhw/X+vXrO/X9AQBA6ArojM5Zx48f1/Hjx/V///d/6tatm3w+X6dfUh4VFaXly5crPT1d1dXVysjI0MSJEzu9cAEAgNAT0BmdL7/8UhMmTNDgwYM1ceJEHT9+XJI0e/bsTr+0vG/fvkpPT5ckJSYmKj4+XidPnuzUDAAAIDQFVHTuuecedenSRUePHlX37t3949OnT9fmzZvbNNe2bds0efJkJSUlyeFwaMOGDa2OsSxLKSkp6tq1q0aOHKmysrJzzrVz5055PB7179+/TRkAAICZAio6b775ppYsWaJ+/fq1GB80aJA+//zzNs3V2NiotLQ0WZZ1zv3r1q2Ty+XSokWLtGvXLqWlpSknJ0e1tbUtjjt58qRmzZqlFStWtO2PAQAAxgpojU5jY2OLMzlnnTx5UjExMW2aKzc3V7m5uefdX1xcrDlz5ig/P1/Sd1d8bdy4UStXrlRhYaEkqampSVOmTFFhYaFuuOGGC75fU1OTmpqa/Ntut7tNeQEAwOUjoDM6Y8eO1Zo1a/zbDodDXq9XS5cu1fjx44MWrrm5WTt37lR2drZ/LCIiQtnZ2dq+fbskyefz6c4779SNN96oO+6446JzFhUVKS4uzv/iZy4AAMwV0BmdpUuXasKECdqxY4eam5t1//3363//+59Onjyp999/P2jhTpw4IY/Ho4SEhBbjCQkJ2rdvnyTp/fff17p16zR8+HD/+p6//e1vGjZs2DnnXLBgQYunr7vdbsoOAACGCqjoDB06VAcOHNBTTz2l2NhYNTQ0aNq0aXI6nerbt2+wM17QmDFj5PV6L/n4mJiYNv+8BgAALk9tLjpnzpzRzTffrJKSEj3wwAMdkckvPj5ekZGRqqmpaTFeU1OjxMTEds1tWZYsy5LH42nXPAAAIHS1eY1Oly5d9PHHH3dEllaio6OVkZGh0tJS/5jX61VpaalGjRrVrrmdTqf27t2r8vLy9sYEAAAhKqDFyDNnztTzzz8flAANDQ2qqKhQRUWFJOnw4cOqqKjQ0aNHJUkul0vPPfecXnjhBX3yySf63e9+p8bGRv9VWAAAAOcT0Bqdb7/9VitXrtRbb72ljIyMVo9bKC4uvuS5duzY0eJKrbMLhfPy8rR69WpNnz5ddXV1Wrhwoaqrq5Wenq7Nmze3WqAMAADwfW0qOp999plSUlK0Z88eXX/99ZKkAwcOtDimrc+6GjdunHw+3wWPKSgoUEFBQZvmvZhwWKOTUrjR7ggAANiqTUVn0KBBOn78uLZu3Srpu0c+PPHEE5fl2RWn0ymn0ym32624uDi74wAAgA7QpjU63z/zsmnTJjU2NgY1EAAAQLAEtBj5rIv95AQAAGCnNhUdh8PRag1OW9fkhArLspSamqoRI0bYHQUAAHSQNq3ROftcqbN3Fj59+rR++9vftrrq6pVXXglewg7CGh0AAMzXpqKTl5fXYnvmzJlBDQMAABBMbSo6q1at6qgcAAAAQdeuxcgAAAChLGyLDouRAQAwX9gWHR7qCQCA+cK26AAAAPNRdAAAgLEoOgAAwFhhW3RYjAwAgPnCtuiwGBkAAPOFbdEBAADmo+gAAABjUXQAAICxKDoAAMBYFB0AAGCssC06XF4OAID5wrbocHk5AADmC9uiAwAAzEfRAQAAxqLoAAAAY1F0AACAsSg6AADAWBQdAABgLIoOAAAwFkUHAAAYK2yLDndGBgDAfGFbdLgzMgAA5gvbogMAAMwXZXcABEdK4Ua7IwAAEHI4owMAAIxF0QEAAMai6AAAAGNRdAAAgLEoOgAAwFgUHQAAYCyKDgAAMBZFBwAAGCtsiw7PugIAwHxhW3R41hUAAOYL26IDAADMR9EBAADGougAAABjUXQAAICxKDoAAMBYFB0AAGAsig4AADAWRQcAABiLogMAAIxF0QEAAMai6AAAAGNRdAAAgLGi7A6AwKQUbrQ7AgAAIY8zOgAAwFhGFJ2pU6eqZ8+e+uUvf2l3FAAAEEKMKDpz587VmjVr7I4BAABCjBFFZ9y4cYqNjbU7BgAACDG2F51t27Zp8uTJSkpKksPh0IYNG1odY1mWUlJS1LVrV40cOVJlZWWdHxQAAFx2bC86jY2NSktLk2VZ59y/bt06uVwuLVq0SLt27VJaWppycnJUW1vbyUkBAMDlxvbLy3Nzc5Wbm3ve/cXFxZozZ47y8/MlSSUlJdq4caNWrlypwsLCNr9fU1OTmpqa/Ntut7vtoQEAwGXB9jM6F9Lc3KydO3cqOzvbPxYREaHs7Gxt3749oDmLiooUFxfnf/Xv3z9YcQEAQIgJ6aJz4sQJeTweJSQktBhPSEhQdXW1fzs7O1u33nqr/v3vf6tfv34XLEELFixQfX29/1VZWdlh+QEAgL1s/+kqGN56661LPjYmJkYxMTEdmAYAAISKkD6jEx8fr8jISNXU1LQYr6mpUWJiYrvmtixLqampGjFiRLvmAQAAoSuki050dLQyMjJUWlrqH/N6vSotLdWoUaPaNbfT6dTevXtVXl7e3pgAACBE2f7TVUNDgw4dOuTfPnz4sCoqKtSrVy8lJyfL5XIpLy9PmZmZysrK0vLly9XY2Oi/CgsAAOB8bC86O3bs0Pjx4/3bLpdLkpSXl6fVq1dr+vTpqqur08KFC1VdXa309HRt3ry51QJlAACA77O96IwbN04+n++CxxQUFKigoCCo72tZlizLksfjCeq8AAAgdIT0Gp2OxBodAADMF7ZFBwAAmI+iAwAAjBW2RYf76AAAYL6wLTqs0QEAwHxhW3QAAID5KDoAAMBYFB0AAGCssC06LEYGAMB8YVt0WIwMAID5wrboAAAA81F0AACAsSg6AADAWBQdAABgrCi7A9jFsixZliWPx9Nh75FSuLHV2JHFk9r87y7l3wAAgNbC9owOV10BAGC+sC06AADAfBQdAABgLIoOAAAwFkUHAAAYi6IDAACMFbZFh4d6AgBgvrAtOlxeDgCA+cK26AAAAPNRdAAAgLEoOgAAwFgUHQAAYCyKDgAAMBZFBwAAGIuiAwAAjBW2RYcbBgIAYL6wLTrcMBAAAPOFbdEBAADmo+gAAABjUXQAAICxKDoAAMBYFB0AAGAsig4AADAWRQcAABiLogMAAIxF0QEAAMai6AAAAGNRdAAAgLGi7A5gF8uyZFmWPB5Pp75vSuHGFttHFk9q878BAACXJmzP6PBQTwAAzBe2RQcAAJiPogMAAIxF0QEAAMai6AAAAGNRdAAAgLEoOgAAwFgUHQAAYCyKDgAAMBZFBwAAGIuiAwAAjEXRAQAAxqLoAAAAY1F0AACAsSg6AADAWEYUnTfeeEPXXHONBg0apL/+9a92xwEAACEiyu4A7fXtt9/K5XJp69atiouLU0ZGhqZOnarevXvbHQ0AANjssj+jU1ZWpiFDhuiqq67SD37wA+Xm5urNN9+0OxYAAAgBthedbdu2afLkyUpKSpLD4dCGDRtaHWNZllJSUtS1a1eNHDlSZWVl/n1VVVW66qqr/NtXXXWVjh071hnRAQBAiLO96DQ2NiotLU2WZZ1z/7p16+RyubRo0SLt2rVLaWlpysnJUW1tbUDv19TUJLfb3eIFAADMZPsandzcXOXm5p53f3FxsebMmaP8/HxJUklJiTZu3KiVK1eqsLBQSUlJLc7gHDt2TFlZWeedr6ioSH/+85+D9we0U0rhRrsjAABgLNvP6FxIc3Ozdu7cqezsbP9YRESEsrOztX37dklSVlaW9uzZo2PHjqmhoUGbNm1STk7OeedcsGCB6uvr/a/KysoO/zsAAIA9bD+jcyEnTpyQx+NRQkJCi/GEhATt27dPkhQVFaVly5Zp/Pjx8nq9uv/++y94xVVMTIxiYmI6NDcAAAgNIV10LtUtt9yiW265xe4YAAAgxIT0T1fx8fGKjIxUTU1Ni/GamholJia2a27LspSamqoRI0a0ax4AABC6QrroREdHKyMjQ6Wlpf4xr9er0tJSjRo1ql1zO51O7d27V+Xl5e2NCQAAQpTtP101NDTo0KFD/u3Dhw+roqJCvXr1UnJyslwul/Ly8pSZmamsrCwtX75cjY2N/quwAAAAzsf2orNjxw6NHz/ev+1yuSRJeXl5Wr16taZPn666ujotXLhQ1dXVSk9P1+bNm1stUG4ry7JkWZY8Hk+75gEAAKHL4fP5fHaHsJPb7VZcXJzq6+vVo0ePoM59Odwj58jiSXZHAAAY5PvffR31PXOp398hvUYHAACgPSg6AADAWBQdAABgrLAtOtxHBwAA84Vt0eE+OgAAmC9siw4AADAfRQcAABjL9hsG2u3sbYTcbnfQ5/Y2fRP0OYOtI/5uAED4+v53X0d9z5yd92K3AwzbGwaevTNyc3OzPv30U7vjAACAAFRWVqpfv37n3R+2Recsr9erqqoqxcbGyuFwXPK/c7vd6t+/vyorK4N+R2W0xGfdufi8Oxefd+fi8+5cHfl5+3w+nTp1SklJSYqIOP9KnLD/6SoiIuKCTfBievTowf9ZOgmfdefi8+5cfN6di8+7c3XU5x0XF3fRY1iMDAAAjEXRAQAAxqLoBCgmJkaLFi1STEyM3VGMx2fdufi8Oxefd+fi8+5cofB5h/1iZAAAYC7O6AAAAGNRdAAAgLEoOgAAwFgUHQAAYCyKTgAsy1JKSoq6du2qkSNHqqyszO5IRioqKtKIESMUGxurPn36aMqUKdq/f7/dscLG4sWL5XA4NG/ePLujGOvYsWOaOXOmevfurW7dumnYsGHasWOH3bGM5PF49OCDD2rgwIHq1q2bfvzjH+uhhx666HOScGm2bdumyZMnKykpSQ6HQxs2bGix3+fzaeHCherbt6+6deum7OxsHTx4sFOyUXTaaN26dXK5XFq0aJF27dqltLQ05eTkqLa21u5oxnnnnXfkdDr1wQcfaMuWLTpz5oxuuukmNTY22h3NeOXl5Xr22Wc1fPhwu6MY66uvvtLo0aPVpUsXbdq0SXv37tWyZcvUs2dPu6MZacmSJXrmmWf01FNP6ZNPPtGSJUu0dOlSPfnkk3ZHM0JjY6PS0tJkWdY59y9dulRPPPGESkpK9OGHH+qKK65QTk6OTp8+3fHhfGiTrKwsn9Pp9G97PB5fUlKSr6ioyMZU4aG2ttYnyffOO+/YHcVop06d8g0aNMi3ZcsW389+9jPf3Llz7Y5kpD/+8Y++MWPG2B0jbEyaNMl31113tRibNm2ab8aMGTYlMpck36uvvurf9nq9vsTERN9jjz3mH/v66699MTExvpdeeqnD83BGpw2am5u1c+dOZWdn+8ciIiKUnZ2t7du325gsPNTX10uSevXqZXMSszmdTk2aNKnF/84RfK+//royMzN16623qk+fPrruuuv03HPP2R3LWDfccINKS0t14MABSdLu3bv13nvvKTc31+Zk5jt8+LCqq6tb/DclLi5OI0eO7JTvzrB/qGdbnDhxQh6PRwkJCS3GExIStG/fPptShQev16t58+Zp9OjRGjp0qN1xjLV27Vrt2rVL5eXldkcx3meffaZnnnlGLpdLf/rTn1ReXq67775b0dHRysvLszuecQoLC+V2u3XttdcqMjJSHo9HjzzyiGbMmGF3NONVV1dL0jm/O8/u60gUHVwWnE6n9uzZo/fee8/uKMaqrKzU3LlztWXLFnXt2tXuOMbzer3KzMzUo48+Kkm67rrrtGfPHpWUlFB0OsDLL7+sf/zjH3rxxRc1ZMgQVVRUaN68eUpKSuLzNhw/XbVBfHy8IiMjVVNT02K8pqZGiYmJNqUyX0FBgd544w1t3bpV/fr1szuOsXbu3Kna2lpdf/31ioqKUlRUlN555x098cQTioqKksfjsTuiUfr27avU1NQWYz/5yU909OhRmxKZ7b777lNhYaFuu+02DRs2THfccYfuueceFRUV2R3NeGe/H+367qTotEF0dLQyMjJUWlrqH/N6vSotLdWoUaNsTGYmn8+ngoICvfrqq3r77bc1cOBAuyMZbcKECfrvf/+riooK/yszM1MzZsxQRUWFIiMj7Y5olNGjR7e6XcKBAwc0YMAAmxKZ7ZtvvlFERMuvvMjISHm9XpsShY+BAwcqMTGxxXen2+3Whx9+2Cnfnfx01UYul0t5eXnKzMxUVlaWli9frsbGRuXn59sdzThOp1MvvviiXnvtNcXGxvp/y42Li1O3bt1sTmee2NjYVuufrrjiCvXu3Zt1UR3gnnvu0Q033KBHH31Uv/rVr1RWVqYVK1ZoxYoVdkcz0uTJk/XII48oOTlZQ4YM0UcffaTi4mLddddddkczQkNDgw4dOuTfPnz4sCoqKtSrVy8lJydr3rx5evjhhzVo0CANHDhQDz74oJKSkjRlypSOD9fh13UZ6Mknn/QlJyf7oqOjfVlZWb4PPvjA7khGknTO16pVq+yOFja4vLxj/etf//INHTrUFxMT47v22mt9K1assDuSsdxut2/u3Lm+5ORkX9euXX0/+tGPfA888ICvqanJ7mhG2Lp16zn/e52Xl+fz+b67xPzBBx/0JSQk+GJiYnwTJkzw7d+/v1OyOXw+bgsJAADMxBodAABgLIoOAAAwFkUHAAAYi6IDAACMRdEBAADGougAAABjUXQAAICxKDoAAMBYFB0AAGAsig4AADAWRQcAABiLogMAAIz1/xVNCkiqXTYxAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ],
      "source": [
        "w = gapfill_higher_def_cost.weights\n",
        "pd.Series(w).plot.hist(bins=100, label='Higher default cost', logy=True)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "XJ-S1mBecSAa",
        "outputId": "f1c85673-fac6-4ba1-d7f4-0d038429e90e",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  79156 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39578 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19789 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9894 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4947 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2473 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1236 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  618 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  309 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  154 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  77 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  5 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.020721.\n",
            "Read LP format model from file /tmp/tmp1q4gd6wi.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 784 rows, 1614 columns, 7198 nonzeros\n",
            "NN gapfilling added 81 new reactions\n",
            "The NN gapfilled model, comes with 807 reactions and 784 metabolites\n",
            "Number of reactions added: 81\n"
          ]
        }
      ],
      "source": [
        "gapfill_higher_def_cost.gapfill()\n",
        "print(\"Number of reactions added:\", len(gapfill_higher_def_cost.added_reactions))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nCmUoBzFcSAb"
      },
      "source": [
        "## Gapfilling with a different medium\n",
        "\n",
        "Often organisms dont grow in a complete medium, you can provide the gapfiller with a medium file"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {
        "id": "eEE4ReskcSAb"
      },
      "outputs": [],
      "source": [
        "medium_file_path = 'DNNGIOR/docs/biochemistry/Nitrogen-Nitrite_media.tsv'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {
        "id": "NUTg6SadcSAc",
        "outputId": "8bd0e922-2cf5-44e5-ba72-ad2073a3f853",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 708
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          id          name  concentration  minflux  maxflux\n",
              "0   cpd00027     D-Glucose          0.001     -100        5\n",
              "1   cpd00075       Nitrite          0.001     -100        5\n",
              "2   cpd00009     Phosphate          0.001     -100        5\n",
              "3   cpd00048       Sulfate          0.001     -100        5\n",
              "4   cpd00063          Ca2+          0.001     -100      100\n",
              "5   cpd00011           CO2          0.001     -100        0\n",
              "6   cpd10516           fe3          0.001     -100      100\n",
              "7   cpd00067            H+          0.001     -100      100\n",
              "8   cpd00001           H2O          0.001     -100      100\n",
              "9   cpd00205            K+          0.001     -100      100\n",
              "10  cpd00254            Mg          0.001     -100      100\n",
              "11  cpd00971           Na+          0.001     -100      100\n",
              "12  cpd00007            O2          0.001     -100      100\n",
              "13  cpd00099           Cl-          0.001     -100      100\n",
              "14  cpd00058          Cu2+          0.001     -100      100\n",
              "15  cpd00149          Co2+          0.001     -100      100\n",
              "16  cpd00030          Mn2+          0.001     -100      100\n",
              "17  cpd00034          Zn2+          0.001     -100      100\n",
              "18  cpd10515          Fe2+          0.001     -100      100\n",
              "19  cpd15432        LipidA          0.001     -100      100\n",
              "20  cpd15560  Ubiquinone-8          0.001     -100      100"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b7c531cd-5585-49b6-9188-236fa7205927\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>id</th>\n",
              "      <th>name</th>\n",
              "      <th>concentration</th>\n",
              "      <th>minflux</th>\n",
              "      <th>maxflux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>cpd00027</td>\n",
              "      <td>D-Glucose</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>cpd00075</td>\n",
              "      <td>Nitrite</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>cpd00009</td>\n",
              "      <td>Phosphate</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>cpd00048</td>\n",
              "      <td>Sulfate</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>cpd00063</td>\n",
              "      <td>Ca2+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>cpd00011</td>\n",
              "      <td>CO2</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>cpd10516</td>\n",
              "      <td>fe3</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>cpd00067</td>\n",
              "      <td>H+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>cpd00001</td>\n",
              "      <td>H2O</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>cpd00205</td>\n",
              "      <td>K+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>cpd00254</td>\n",
              "      <td>Mg</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>cpd00971</td>\n",
              "      <td>Na+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>cpd00007</td>\n",
              "      <td>O2</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>cpd00099</td>\n",
              "      <td>Cl-</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>cpd00058</td>\n",
              "      <td>Cu2+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>cpd00149</td>\n",
              "      <td>Co2+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>cpd00030</td>\n",
              "      <td>Mn2+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>cpd00034</td>\n",
              "      <td>Zn2+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>cpd10515</td>\n",
              "      <td>Fe2+</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>cpd15432</td>\n",
              "      <td>LipidA</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>cpd15560</td>\n",
              "      <td>Ubiquinone-8</td>\n",
              "      <td>0.001</td>\n",
              "      <td>-100</td>\n",
              "      <td>100</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b7c531cd-5585-49b6-9188-236fa7205927')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b7c531cd-5585-49b6-9188-236fa7205927 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b7c531cd-5585-49b6-9188-236fa7205927');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-8887226b-52d7-473f-9029-37d6d59c8990\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-8887226b-52d7-473f-9029-37d6d59c8990')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-8887226b-52d7-473f-9029-37d6d59c8990 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"pd\",\n  \"rows\": 21,\n  \"fields\": [\n    {\n      \"column\": \"id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"cpd00027\",\n          \"cpd00034\",\n          \"cpd00149\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"D-Glucose\",\n          \"Zn2+\",\n          \"Co2+\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"concentration\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 4.4439065127221605e-19,\n        \"min\": 0.001,\n        \"max\": 0.001,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          0.001\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"minflux\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 0,\n        \"min\": -100,\n        \"max\": -100,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          -100\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"maxflux\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 41,\n        \"min\": 0,\n        \"max\": 100,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          5\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ],
      "source": [
        "pd.read_csv(medium_file_path, sep='\\t')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "id": "uwR5w-lRcSAc",
        "outputId": "5987aef3-09e8-43eb-81cf-5efe6f87cffe",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Loading medium from: DNNGIOR/docs/biochemistry/Nitrogen-Nitrite_media.tsv\n",
            "#reactions not found in NN-keys:  4 / 665\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  76914 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38457 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19228 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9614 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4807 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2403 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1201 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  600 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  300 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  150 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  75 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  37 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  18 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.344953.\n",
            "Read LP format model from file /tmp/tmpcuf9nivt.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 865 rows, 1730 columns, 8410 nonzeros\n",
            "NN gapfilling added 200 new reactions\n",
            "The NN gapfilled model, comes with 865 reactions and 865 metabolites\n"
          ]
        }
      ],
      "source": [
        "gapfill_nit_medium = Gapfill(draftModel = path_to_draft_model, objectiveName = 'bio1', medium_file = medium_file_path)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "id": "ysWxbk5bcSAd",
        "outputId": "ee366d18-306a-47ac-ff29-80d79a111823",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 849
        }
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<cobra.summary.model_summary.ModelSummary at 0x79b11a7141d0>"
            ],
            "text/html": [
              "<h3>Objective</h3><p>1.0 bio1 = 46.35275895999998</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>cpd00030_e0</td>\n",
              "      <td>EX_cpd00030_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00034_e0</td>\n",
              "      <td>EX_cpd00034_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00058_e0</td>\n",
              "      <td>EX_cpd00058_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00063_e0</td>\n",
              "      <td>EX_cpd00063_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00099_e0</td>\n",
              "      <td>EX_cpd00099_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00205_e0</td>\n",
              "      <td>EX_cpd00205_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00254_e0</td>\n",
              "      <td>EX_cpd00254_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd10515_e0</td>\n",
              "      <td>EX_cpd10515_e0</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15560_e0</td>\n",
              "      <td>EX_cpd15560_e0</td>\n",
              "      <td>0.3333</td>\n",
              "      <td>49</td>\n",
              "      <td>100.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd17041_c0</td>\n",
              "      <td>rxn13782_c0</td>\n",
              "      <td>46.35</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd17042_c0</td>\n",
              "      <td>rxn13783_c0</td>\n",
              "      <td>46.35</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd17043_c0</td>\n",
              "      <td>rxn13784_c0</td>\n",
              "      <td>46.35</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>cpd00066_e0</td>\n",
              "      <td>EX_cpd00066_e0</td>\n",
              "      <td>-0.1763</td>\n",
              "      <td>9</td>\n",
              "      <td>16.27%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00119_e0</td>\n",
              "      <td>EX_cpd00119_e0</td>\n",
              "      <td>-0.4286</td>\n",
              "      <td>6</td>\n",
              "      <td>26.38%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00129_e0</td>\n",
              "      <td>EX_cpd00129_e0</td>\n",
              "      <td>-0.4646</td>\n",
              "      <td>5</td>\n",
              "      <td>23.83%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00149_e0</td>\n",
              "      <td>EX_cpd00149_e0</td>\n",
              "      <td>-0.3333</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd00508_e0</td>\n",
              "      <td>EX_cpd00508_e0</td>\n",
              "      <td>-0.1025</td>\n",
              "      <td>6</td>\n",
              "      <td>6.31%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd01981_e0</td>\n",
              "      <td>EX_cpd01981_e0</td>\n",
              "      <td>-0.05295</td>\n",
              "      <td>6</td>\n",
              "      <td>3.26%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd11416_c0</td>\n",
              "      <td>SK_cpd11416_c0</td>\n",
              "      <td>-46.35</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cpd15378_c0</td>\n",
              "      <td>rxn30282_c0</td>\n",
              "      <td>-0.3333</td>\n",
              "      <td>7</td>\n",
              "      <td>23.94%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>"
            ]
          },
          "metadata": {},
          "execution_count": 19
        }
      ],
      "source": [
        "gapfill_nit_medium.gapfilledModel.summary()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5bYYy6p8cSAd"
      },
      "source": [
        "Relevant to note is that using a more minimal medium tends to lead to a more gapfilled reactions as the models have to synthesize more compounds themselves"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VYDLbzLCcSAe",
        "outputId": "2f0ac4b6-c1bf-40ee-dd0d-22302f20a8c7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Number of reactions added: 198\n"
          ]
        }
      ],
      "source": [
        "print(\"Number of reactions added:\", len(gapfill_nit_medium.added_reactions))"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "D_B9V-zscSAe"
      },
      "source": [
        "### Batch gapfilling using the command line interface (CLI)\n",
        "\n",
        "Finally in the case where you have a lot of models you want to gapfill with the same medium you can use the CLI\n",
        "\n",
        "`python fasta2model_CLI.py -f DIR_FASTA -o output_folder`\n",
        "\n",
        "This command will create an output folder (-o) containing a subfolder with base ungapfilled models, a subfolder with gapfilled models, a log, and a tsv file telling you the number of added reactions.\n",
        "\n",
        "This CLI has limited functionality and assumes the same conditions for all gapfilling but you can change the standard gapfilling medium using the -e parameter.\n",
        "\n",
        "`python fasta2model_CLI.py -f DIR_FASTA -o DIR_OUTPUT' -e PATH_TO_MEDIUM_FILE`\n",
        "\n",
        "if you allready have base models you can use the -m parameter to provide a folder with base models to skip the base model building step.\n",
        "\n",
        "`python fasta2model_CLI.py -m DIR_MODELS -o DIR_OUTPUT`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 20,
      "metadata": {
        "collapsed": true,
        "id": "BmCpcFg9cSAf",
        "outputId": "9b49d2f2-42ff-4e81-f11a-99778fd0497a",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# Creating output_folder for gapfilled models\n",
            "Set parameter WLSAccessID\n",
            "Set parameter WLSSecret\n",
            "Set parameter LicenseID to value 940603\n",
            "Academic license 940603 - for non-commercial use only - registered to da___@gmail.com\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1741265521.533081    5908 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1741265521.550368    5908 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "WARNING: To enable the NN_Trainer script, you need to install tensorflow <https://www.tensorflow.org/install>\n",
            "The rest of dnngior features can be used without it.\n",
            "Gap-filling database =  ModelSEED\n",
            "Loading medium from: ./DNNGIOR/docs/biochemistry/Nitrogen-Nitrite_media.tsv\n",
            "#reactions not found in NN-keys:  7 / 648\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  76950 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38475 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19237 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9618 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4809 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2404 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1202 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  601 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  300 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  150 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  75 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  37 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  18 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.517733.\n",
            "Read LP format model from file /tmp/tmpupnys0jl.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 892 rows, 1644 columns, 7866 nonzeros\n",
            "NN gapfilling added 174 new reactions\n",
            "The NN gapfilled model, comes with 822 reactions and 892 metabolites\n",
            "Gap-filling database =  ModelSEED\n",
            "Loading medium from: ./DNNGIOR/docs/biochemistry/Nitrogen-Nitrite_media.tsv\n",
            "#reactions not found in NN-keys:  6 / 913\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  76525 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38262 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19131 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9565 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4782 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2391 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1195 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  597 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  298 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  149 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  74 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  37 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  18 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.000000.\n",
            "Read LP format model from file /tmp/tmp6l3i60dd.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 1037 rows, 2074 columns, 10042 nonzeros\n",
            "NN gapfilling added 124 new reactions\n",
            "The NN gapfilled model, comes with 1037 reactions and 1037 metabolites\n",
            "Gap-filling database =  ModelSEED\n",
            "Loading medium from: ./DNNGIOR/docs/biochemistry/Nitrogen-Nitrite_media.tsv\n",
            "#reactions not found in NN-keys:  6 / 777\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  76702 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38351 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19175 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9587 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4793 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2396 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1198 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  599 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "Interrupt request received\n",
            "^C\n"
          ]
        }
      ],
      "source": [
        "!python DNNGIOR/dnngior/fasta2model_CLI.py -m one_per_phylum_models -o one_per_phylum_gapfilled -e DNNGIOR/docs/biochemistry/Nitrogen-Nitrite_media.tsv -sm .sbml"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "This will take a while so maybe we should skip this or create a subset or let them continue"
      ],
      "metadata": {
        "id": "c8kJuAA3hrld"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IZWLxxrRcSAf"
      },
      "source": [
        "## For the interested: Customizing the candidate reaction scores\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Pnc-sPRgcSAf"
      },
      "source": [
        "\n",
        "### Changing the Neural network\n",
        "\n",
        "By default, dnngior uses its built-in neural network (NN) to predict which reactions should be added during gap filling. The gap-filler automatically assigns a cost of zero to reactions already present in the draft model, as well as to the exchange reactions defined in the media. For reactions not included in the NN-predicted reactome, a default high cost (set to 1) is applied, as these are considered less likely to occur.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "VOjiKCyXcSAg",
        "outputId": "12b90541-3fa8-4e55-f07a-2e1d94c35eb6"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  BiGG\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  1 / 530\n",
            "of which are exchanges:  0\n",
            "WARNING: reaction St in candidates but not in database, removing from candidates\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "\n",
            "\n",
            " condition is currently:  54652 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  27326 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  13663 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  6831 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  3415 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  1707 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  853 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  426 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  213 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  106 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  53 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  26 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  13 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  7 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  3 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.019842.\n",
            "Read LP format model from file /tmp/tmp72ezq61j.lp\n",
            "Reading time = 0.00 seconds\n",
            ": 737 rows, 1338 columns, 6710 nonzeros\n",
            "skipping refinement, not currently supported for BiGG models\n",
            "NN gapfilling added 139 new reactions\n",
            "The NN gapfilled model, comes with 669 reactions and 737 metabolites\n"
          ]
        }
      ],
      "source": [
        "gapfill_Bigg = Gapfill('DNNGIOR/docs/models/bigg_example.xml', objectiveName='Growth', dbType='BiGG')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "83xq05DqcSAg"
      },
      "source": [
        "The second customization option is to change the neural network (NN) used during gap filling. We will train our own network in the next part of the workshop. For now we can use an example provided, we made a couple neural networks tailored to some of the top phyla in docs/NN/custom_networks/.\n",
        "\n",
        "To specify a custom NN, update the trainedNNPath parameter accordingly."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "X_EgqJfycSAh"
      },
      "outputs": [],
      "source": [
        "dif_NN_gapfill = Gapfill(path_to_draft_model, trainedNNPath='DNNGIOR/docs/NN/custom_networks/Phylum_Actinobacteria.npz')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "arzhqO8ZcSAh"
      },
      "source": [
        "### Blacklisting reactions\n",
        "\n",
        "You might want to exclude specific reactions from the gap-filling database (e.g. they are unbalanced or you know cannot be present based on other data), this can be done using the blacklist argument:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "x8Wmu9x0cSAi"
      },
      "outputs": [],
      "source": [
        "blackList = ['rxn99999_c0']\n",
        "gapfill_with_blacklist = Gapfill(path_to_draft_model, black_list = blackList, objectiveName = 'bio1')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IlWZ4ixbcSAi"
      },
      "source": [
        "Sometimes reactions are unavoidable (i.e. no solution can be found without them), but you want as few of them as possible. You could manually set the weights of these reactions (see below) but to make it easier you can set a grey list. By default these reactions get a cost of 1,000 but you can change this using punish_cost."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cdvOYMZ-cSAj"
      },
      "outputs": [],
      "source": [
        "greyList = ['rxn04070_c0','rxn05467_c0','rxn00543_c0']\n",
        "gapfill_with_greylist = Gapfill(path_to_draft_model, grey_list = greyList, punish_cost = 5000, objectiveName = 'bio1')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iu2q3Z9zcSAt"
      },
      "source": [
        "### Custom scores\n",
        "\n",
        "The final option is to use fully custom scores. This is by far the most, you can make any changes to any reaction you want. To make these changes it is useful to set the gapfill parameter to False. This stops the gapfilling class from automatically continuing to the gapfilling step."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lAP11_-6cSAu",
        "outputId": "7593c5dc-62d1-49c1-8b00-913d1777e1fe"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "'' is not a valid SBML 'SId'.\n",
            "'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  65 / 726\n",
            "of which are exchanges:  61\n"
          ]
        }
      ],
      "source": [
        "ungapfilled_model = Gapfill(draftModel = path_to_draft_model, gapfill=False)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0SkiXYg8cSAu"
      },
      "source": [
        "Then you can change your candidates using the set_weights function"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ne1huIxecSAu"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "s_file = open(os.path.join('DNNGIOR', 'docs', 'NN','single_example.json'))\n",
        "scores = json.loads(s_file.read())\n",
        "s_file.close()\n",
        "\n",
        "ungapfilled_model.set_weights(scores)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YrFaq-QacSAv"
      },
      "source": [
        "or manually set reactions directly"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "un8IyfyRcSAv"
      },
      "outputs": [],
      "source": [
        "ungapfilled_model.weights['rnx0001'] = 0.4"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cLWh6H42cSAw"
      },
      "source": [
        "reloading your model will reset them back to the NN-predicted weights but there is also a function for this:"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8gYK3NvLcSAw",
        "outputId": "ad0ecbd9-0062-4b00-cc22-38aa714a4329"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  65 / 726\n",
            "of which are exchanges:  61\n"
          ]
        }
      ],
      "source": [
        "ungapfilled_model.reset_weights()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JciCUTPacSAx"
      },
      "source": [
        "Once you are ready to continue the gapfilling, you can use the class function gapfill() to resume the gap-filling process"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "x5HN1LxtcSAx",
        "outputId": "af422d49-054a-4121-e3dc-59e3c5b55d96"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  79817 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39908 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19954 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9977 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4988 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2494 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1247 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  623 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  311 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  155 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  77 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.392904.\n",
            "Read LP format model from file /tmp/tmpqpqfmmna.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 801 rows, 1662 columns, 7428 nonzeros\n",
            "NN gapfilling added 105 new reactions\n",
            "The NN gapfilled model, comes with 831 reactions and 801 metabolites\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Name</strong></td>\n",
              "                <td>_gapfilled</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>7feb32abd330</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Number of metabolites</strong></td>\n",
              "                <td>801</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Number of reactions</strong></td>\n",
              "                <td>831</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Number of genes</strong></td>\n",
              "                <td>529</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Number of groups</strong></td>\n",
              "                <td>0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Objective expression</strong></td>\n",
              "                <td>1.0*bio1 - 1.0*bio1_reverse_b18f7</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartments</strong></td>\n",
              "                <td>c0, e0</td>\n",
              "            </tr>\n",
              "          </table>"
            ],
            "text/plain": [
              "<Model _gapfilled at 0x7feb32abd330>"
            ]
          },
          "execution_count": 106,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "ungapfilled_model.gapfill()"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "tfcob",
      "language": "python",
      "name": "tfcob"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.14"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}