{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "colab_type": "text",
        "id": "view-in-github"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/hariszaf/metabolic_toy_model/blob/main/Antony2025/participantProjects.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aaqbLrQy5894"
      },
      "source": [
        "# **Finding Optimal Ecological Partners For Gut Colonization** (Julian Tap / Lu Chen)\n",
        "\n",
        "\\\\\n",
        "\n",
        "# **Impact of methanol on anaerobic digestion** (Ariane Bize)\n",
        "\n",
        "\\\\\n",
        "\n",
        "# **Understanding the communities in the anode and cathode of bioelectrosynthesis reactors** (Louise Rigaud / Wenbo Sui)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mCTvF-KZ-PSJ"
      },
      "source": [
        "### **Basic Setup**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FJ1DXzsd-eeW"
      },
      "source": [
        "Setup **Gurobi**, **COBRApy**, **modelSEEDpy**, and **DNNGIOR**. See the [setting up your environment](https://colab.research.google.com/github/hariszaf/metabolic_toy_model/blob/main/Antony2025/preparingYourEnvironment.ipynb)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kOPIsNzZ55yw",
        "outputId": "33c3b376-ada5-4700-8ebd-e4906e016ed4"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "License file created at /content/licenses/gurobi.lic\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "def create_gurobi_license():\n",
        "    license_content = (\n",
        "        \"# Gurobi WLS license file\\n\"\n",
        "        \"# Your credentials are private and should not be shared or copied to public repositories.\\n\"\n",
        "        \"# Visit https://license.gurobi.com/manager/doc/overview for more information.\\n\"\n",
        "        \"WLSACCESSID=1fedf73b-9471-4da8-bdc7-2aaacf2e30f3\\n\"\n",
        "        \"WLSSECRET=3bc7d209-a4ec-4195-98be-4b254f181512\\n\"\n",
        "        \"LICENSEID=940603\"\n",
        "    )\n",
        "    with open(\"/content/licenses/gurobi.lic\", \"w\") as f:\n",
        "        f.write(license_content)\n",
        "    print(\"License file created at /content/licenses/gurobi.lic\")\n",
        "\n",
        "\n",
        "\n",
        "# Create directory for the license\n",
        "os.makedirs(\"/content/licenses\", exist_ok=True)\n",
        "\n",
        "# Generate the license file\n",
        "create_gurobi_license()\n",
        "\n",
        "#add to path\n",
        "os.environ['GRB_LICENSE_FILE'] = '/content/licenses/gurobi.lic'"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "m-XCQYhH-rL3",
        "outputId": "8064c9e7-14f5-4fb4-f86e-7ebea28b0ddb"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Collecting gurobipy\n",
            "  Downloading gurobipy-12.0.1-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (16 kB)\n",
            "Downloading gurobipy-12.0.1-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (14.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m14.4/14.4 MB\u001b[0m \u001b[31m49.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: gurobipy\n",
            "Successfully installed gurobipy-12.0.1\n",
            "Collecting cobra\n",
            "  Downloading cobra-0.29.1-py2.py3-none-any.whl.metadata (9.3 kB)\n",
            "Collecting appdirs~=1.4 (from cobra)\n",
            "  Downloading appdirs-1.4.4-py2.py3-none-any.whl.metadata (9.0 kB)\n",
            "Collecting depinfo~=2.2 (from cobra)\n",
            "  Downloading depinfo-2.2.0-py3-none-any.whl.metadata (3.8 kB)\n",
            "Collecting diskcache~=5.0 (from cobra)\n",
            "  Downloading diskcache-5.6.3-py3-none-any.whl.metadata (20 kB)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.11/dist-packages (from cobra) (1.0.0)\n",
            "Requirement already satisfied: httpx~=0.24 in /usr/local/lib/python3.11/dist-packages (from cobra) (0.28.1)\n",
            "Requirement already satisfied: importlib-resources in /usr/local/lib/python3.11/dist-packages (from cobra) (6.5.2)\n",
            "Requirement already satisfied: numpy>=1.13 in /usr/local/lib/python3.11/dist-packages (from cobra) (1.26.4)\n",
            "Collecting optlang~=1.8 (from cobra)\n",
            "  Downloading optlang-1.8.3-py2.py3-none-any.whl.metadata (8.2 kB)\n",
            "Requirement already satisfied: pandas<3.0,>=1.0 in /usr/local/lib/python3.11/dist-packages (from cobra) (2.2.2)\n",
            "Requirement already satisfied: pydantic>=1.6 in /usr/local/lib/python3.11/dist-packages (from cobra) (2.10.6)\n",
            "Collecting python-libsbml~=5.19 (from cobra)\n",
            "  Downloading python_libsbml-5.20.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (532 bytes)\n",
            "Requirement already satisfied: rich>=8.0 in /usr/local/lib/python3.11/dist-packages (from cobra) (13.9.4)\n",
            "Collecting ruamel.yaml~=0.16 (from cobra)\n",
            "  Downloading ruamel.yaml-0.18.10-py3-none-any.whl.metadata (23 kB)\n",
            "Collecting swiglpk (from cobra)\n",
            "  Downloading swiglpk-5.0.12-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.5 kB)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (3.7.1)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (2025.1.31)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (1.0.7)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra) (3.10)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.11/dist-packages (from httpcore==1.*->httpx~=0.24->cobra) (0.14.0)\n",
            "Requirement already satisfied: sympy>=1.12.0 in /usr/local/lib/python3.11/dist-packages (from optlang~=1.8->cobra) (1.13.1)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0,>=1.0->cobra) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0,>=1.0->cobra) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0,>=1.0->cobra) (2025.1)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.2 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra) (2.27.2)\n",
            "Requirement already satisfied: typing-extensions>=4.12.2 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra) (4.12.2)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich>=8.0->cobra) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich>=8.0->cobra) (2.18.0)\n",
            "Collecting ruamel.yaml.clib>=0.2.7 (from ruamel.yaml~=0.16->cobra)\n",
            "  Downloading ruamel.yaml.clib-0.2.12-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.7 kB)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich>=8.0->cobra) (0.1.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas<3.0,>=1.0->cobra) (1.17.0)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy>=1.12.0->optlang~=1.8->cobra) (1.3.0)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio->httpx~=0.24->cobra) (1.3.1)\n",
            "Downloading cobra-0.29.1-py2.py3-none-any.whl (1.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m39.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
            "Downloading depinfo-2.2.0-py3-none-any.whl (24 kB)\n",
            "Downloading diskcache-5.6.3-py3-none-any.whl (45 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m45.5/45.5 kB\u001b[0m \u001b[31m3.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading optlang-1.8.3-py2.py3-none-any.whl (141 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m141.8/141.8 kB\u001b[0m \u001b[31m10.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_libsbml-5.20.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.0/8.0 MB\u001b[0m \u001b[31m58.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ruamel.yaml-0.18.10-py3-none-any.whl (117 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m117.7/117.7 kB\u001b[0m \u001b[31m8.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading swiglpk-5.0.12-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.3/2.3 MB\u001b[0m \u001b[31m58.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ruamel.yaml.clib-0.2.12-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (739 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m739.1/739.1 kB\u001b[0m \u001b[31m33.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: swiglpk, python-libsbml, appdirs, ruamel.yaml.clib, diskcache, depinfo, ruamel.yaml, optlang, cobra\n",
            "Successfully installed appdirs-1.4.4 cobra-0.29.1 depinfo-2.2.0 diskcache-5.6.3 optlang-1.8.3 python-libsbml-5.20.4 ruamel.yaml-0.18.10 ruamel.yaml.clib-0.2.12 swiglpk-5.0.12\n",
            "Collecting pyrodigal\n",
            "  Downloading pyrodigal-3.6.3.post1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (56 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.4/56.4 kB\u001b[0m \u001b[31m586.4 kB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting archspec~=0.2.0 (from pyrodigal)\n",
            "  Downloading archspec-0.2.5-py3-none-any.whl.metadata (4.4 kB)\n",
            "Downloading pyrodigal-3.6.3.post1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.0/3.0 MB\u001b[0m \u001b[31m24.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading archspec-0.2.5-py3-none-any.whl (76 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m76.2/76.2 kB\u001b[0m \u001b[31m4.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: archspec, pyrodigal\n",
            "Successfully installed archspec-0.2.5 pyrodigal-3.6.3.post1\n",
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (1.26.4)\n",
            "Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m60.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.85\n",
            "Collecting dnngior\n",
            "  Downloading dnngior-0.0.5-py3-none-any.whl.metadata (2.8 kB)\n",
            "Downloading dnngior-0.0.5-py3-none-any.whl (29.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m29.2/29.2 MB\u001b[0m \u001b[31m59.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: dnngior\n",
            "Successfully installed dnngior-0.0.5\n",
            "Cloning into 'ModelSEEDpy'...\n",
            "remote: Enumerating objects: 4227, done.\u001b[K\n",
            "remote: Counting objects: 100% (1366/1366), done.\u001b[K\n",
            "remote: Compressing objects: 100% (263/263), done.\u001b[K\n",
            "remote: Total 4227 (delta 1199), reused 1112 (delta 1103), pack-reused 2861 (from 2)\u001b[K\n",
            "Receiving objects: 100% (4227/4227), 8.44 MiB | 7.50 MiB/s, done.\n",
            "Resolving deltas: 100% (2970/2970), done.\n",
            "Processing ./ModelSEEDpy\n",
            "  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: networkx>=2.4 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (3.4.2)\n",
            "Requirement already satisfied: cobra>=0.29.0 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (0.29.1)\n",
            "Requirement already satisfied: pandas>=2.2.2 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (2.2.2)\n",
            "Collecting scikit-learn==1.2.0 (from ModelSEEDpy==0.4.0)\n",
            "  Downloading scikit_learn-1.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)\n",
            "Requirement already satisfied: scipy>=1.5.4 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (1.14.1)\n",
            "Collecting chemicals>=1.0.13 (from ModelSEEDpy==0.4.0)\n",
            "  Downloading chemicals-1.3.2-py3-none-any.whl.metadata (12 kB)\n",
            "Collecting chemw>=0.3.2 (from ModelSEEDpy==0.4.0)\n",
            "  Downloading ChemW-0.3.5.tar.gz (467 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m467.6/467.6 kB\u001b[0m \u001b[31m27.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: matplotlib>=3.0.0 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (3.10.0)\n",
            "Requirement already satisfied: Jinja2>=3.1.4 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (3.1.6)\n",
            "Requirement already satisfied: sympy>=1.12.0 in /usr/local/lib/python3.11/dist-packages (from ModelSEEDpy==0.4.0) (1.13.1)\n",
            "Requirement already satisfied: numpy>=1.17.3 in /usr/local/lib/python3.11/dist-packages (from scikit-learn==1.2.0->ModelSEEDpy==0.4.0) (1.26.4)\n",
            "Requirement already satisfied: joblib>=1.1.1 in /usr/local/lib/python3.11/dist-packages (from scikit-learn==1.2.0->ModelSEEDpy==0.4.0) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from scikit-learn==1.2.0->ModelSEEDpy==0.4.0) (3.5.0)\n",
            "Collecting fluids>=1.1.0 (from chemicals>=1.0.13->ModelSEEDpy==0.4.0)\n",
            "  Downloading fluids-1.1.0-py3-none-any.whl.metadata (7.2 kB)\n",
            "Collecting pubchempy (from chemw>=0.3.2->ModelSEEDpy==0.4.0)\n",
            "  Downloading PubChemPy-1.0.4.tar.gz (29 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from chemw>=0.3.2->ModelSEEDpy==0.4.0) (2.32.3)\n",
            "Collecting sigfig (from chemw>=0.3.2->ModelSEEDpy==0.4.0)\n",
            "  Downloading sigfig-1.3.19-py3-none-any.whl.metadata (4.7 kB)\n",
            "Requirement already satisfied: appdirs~=1.4 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (1.4.4)\n",
            "Requirement already satisfied: depinfo~=2.2 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (2.2.0)\n",
            "Requirement already satisfied: diskcache~=5.0 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (5.6.3)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (1.0.0)\n",
            "Requirement already satisfied: httpx~=0.24 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (0.28.1)\n",
            "Requirement already satisfied: importlib-resources in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (6.5.2)\n",
            "Requirement already satisfied: optlang~=1.8 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (1.8.3)\n",
            "Requirement already satisfied: pydantic>=1.6 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (2.10.6)\n",
            "Requirement already satisfied: python-libsbml~=5.19 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (5.20.4)\n",
            "Requirement already satisfied: rich>=8.0 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (13.9.4)\n",
            "Requirement already satisfied: ruamel.yaml~=0.16 in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (0.18.10)\n",
            "Requirement already satisfied: swiglpk in /usr/local/lib/python3.11/dist-packages (from cobra>=0.29.0->ModelSEEDpy==0.4.0) (5.0.12)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from Jinja2>=3.1.4->ModelSEEDpy==0.4.0) (3.0.2)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (4.56.0)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (1.4.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (3.2.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas>=2.2.2->ModelSEEDpy==0.4.0) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas>=2.2.2->ModelSEEDpy==0.4.0) (2025.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy>=1.12.0->ModelSEEDpy==0.4.0) (1.3.0)\n",
            "Requirement already satisfied: anyio in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra>=0.29.0->ModelSEEDpy==0.4.0) (3.7.1)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra>=0.29.0->ModelSEEDpy==0.4.0) (2025.1.31)\n",
            "Requirement already satisfied: httpcore==1.* in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra>=0.29.0->ModelSEEDpy==0.4.0) (1.0.7)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.11/dist-packages (from httpx~=0.24->cobra>=0.29.0->ModelSEEDpy==0.4.0) (3.10)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.11/dist-packages (from httpcore==1.*->httpx~=0.24->cobra>=0.29.0->ModelSEEDpy==0.4.0) (0.14.0)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra>=0.29.0->ModelSEEDpy==0.4.0) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.2 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra>=0.29.0->ModelSEEDpy==0.4.0) (2.27.2)\n",
            "Requirement already satisfied: typing-extensions>=4.12.2 in /usr/local/lib/python3.11/dist-packages (from pydantic>=1.6->cobra>=0.29.0->ModelSEEDpy==0.4.0) (4.12.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.7->matplotlib>=3.0.0->ModelSEEDpy==0.4.0) (1.17.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich>=8.0->cobra>=0.29.0->ModelSEEDpy==0.4.0) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich>=8.0->cobra>=0.29.0->ModelSEEDpy==0.4.0) (2.18.0)\n",
            "Requirement already satisfied: ruamel.yaml.clib>=0.2.7 in /usr/local/lib/python3.11/dist-packages (from ruamel.yaml~=0.16->cobra>=0.29.0->ModelSEEDpy==0.4.0) (0.2.12)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->chemw>=0.3.2->ModelSEEDpy==0.4.0) (3.4.1)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->chemw>=0.3.2->ModelSEEDpy==0.4.0) (2.3.0)\n",
            "Requirement already satisfied: sortedcontainers<3.0.0,>=2.4.0 in /usr/local/lib/python3.11/dist-packages (from sigfig->chemw>=0.3.2->ModelSEEDpy==0.4.0) (2.4.0)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich>=8.0->cobra>=0.29.0->ModelSEEDpy==0.4.0) (0.1.2)\n",
            "Requirement already satisfied: sniffio>=1.1 in /usr/local/lib/python3.11/dist-packages (from anyio->httpx~=0.24->cobra>=0.29.0->ModelSEEDpy==0.4.0) (1.3.1)\n",
            "Downloading scikit_learn-1.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (9.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m9.5/9.5 MB\u001b[0m \u001b[31m101.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading chemicals-1.3.2-py3-none-any.whl (26.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m26.0/26.0 MB\u001b[0m \u001b[31m20.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading fluids-1.1.0-py3-none-any.whl (627 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m627.5/627.5 kB\u001b[0m \u001b[31m35.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading sigfig-1.3.19-py3-none-any.whl (10 kB)\n",
            "Building wheels for collected packages: ModelSEEDpy, chemw, pubchempy\n",
            "  Building wheel for ModelSEEDpy (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for ModelSEEDpy: filename=modelseedpy-0.4.0-py3-none-any.whl size=175140 sha256=4f0eac1b92545f8b88c6379fc45f026b1480822cbf1d55f27f74854feee9db84\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-_gyst__5/wheels/61/dd/46/ddba7fccc009e806714e45442162eda574b56d2017871b7082\n",
            "  Building wheel for chemw (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for chemw: filename=ChemW-0.3.5-py3-none-any.whl size=480096 sha256=2ad4b800572d4579eca4d2831eab0f9474541ddcaad96cc7c72611ce49fe682b\n",
            "  Stored in directory: /root/.cache/pip/wheels/bf/ab/52/90a06eff7d43377b19bfda55243ac0bbabaa088d4cddd8efcc\n",
            "  Building wheel for pubchempy (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pubchempy: filename=PubChemPy-1.0.4-py3-none-any.whl size=13819 sha256=e03c942147539f0a954aa39a32191a5a04a5711f077aab18b24e46201cd8a983\n",
            "  Stored in directory: /root/.cache/pip/wheels/8b/e3/6c/3385b2db08b0985a87f5b117f98d0cb61a3ae3ca3bcbbd8307\n",
            "Successfully built ModelSEEDpy chemw pubchempy\n",
            "Installing collected packages: pubchempy, sigfig, scikit-learn, fluids, chemicals, chemw, ModelSEEDpy\n",
            "  Attempting uninstall: scikit-learn\n",
            "    Found existing installation: scikit-learn 1.6.1\n",
            "    Uninstalling scikit-learn-1.6.1:\n",
            "      Successfully uninstalled scikit-learn-1.6.1\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "mlxtend 0.23.4 requires scikit-learn>=1.3.1, but you have scikit-learn 1.2.0 which is incompatible.\n",
            "imbalanced-learn 0.13.0 requires scikit-learn<2,>=1.3.2, but you have scikit-learn 1.2.0 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed ModelSEEDpy-0.4.0 chemicals-1.3.2 chemw-0.3.5 fluids-1.1.0 pubchempy-1.0.4 scikit-learn-1.2.0 sigfig-1.3.19\n",
            "Cloning into 'metabolic_toy_model'...\n",
            "remote: Enumerating objects: 717, done.\u001b[K\n",
            "remote: Counting objects: 100% (78/78), done.\u001b[K\n",
            "remote: Compressing objects: 100% (61/61), done.\u001b[K\n",
            "remote: Total 717 (delta 42), reused 40 (delta 17), pack-reused 639 (from 1)\u001b[K\n",
            "Receiving objects: 100% (717/717), 61.81 MiB | 15.41 MiB/s, done.\n",
            "Resolving deltas: 100% (408/408), done.\n",
            "Updating files: 100% (182/182), done.\n"
          ]
        }
      ],
      "source": [
        "# @title\n",
        "!pip install gurobipy\n",
        "!pip install cobra\n",
        "\n",
        "!pip install pyrodigal\n",
        "!pip install biopython\n",
        "\n",
        "!pip install dnngior --no-deps\n",
        "\n",
        "!git clone https://github.com/ModelSEED/ModelSEEDpy\n",
        "!pip install ModelSEEDpy/.\n",
        "\n",
        "!git clone https://github.com/hariszaf/metabolic_toy_model.git"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Fnf28QKU-wff",
        "outputId": "30a2bd7b-a301-4e99-de22-acfb78467305"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Set parameter WLSAccessID\n",
            "Set parameter WLSSecret\n",
            "Set parameter LicenseID to value 940603\n",
            "Academic license 940603 - for non-commercial use only - registered to da___@gmail.com\n",
            "Gurobi is working! 😀\n",
            "flux balance analysis solution is 0.8739215069684305\n",
            "COBRApy is working 😀\n"
          ]
        }
      ],
      "source": [
        "# @title\n",
        "import gurobipy\n",
        "from gurobipy import Model\n",
        "model = Model(\"test\")\n",
        "print(\"Gurobi is working!\", \"\\U0001F600\")\n",
        "\n",
        "#install COBRApy\n",
        "import cobra\n",
        "from cobra.io import load_model\n",
        "model = load_model(\"textbook\")\n",
        "solution = model.optimize()\n",
        "print(f\"flux balance analysis solution is {solution.objective_value}\")\n",
        "print(\"COBRApy is working\", \"\\U0001F600\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uVLQuLz_PDlZ"
      },
      "source": [
        "### **Import all the packages we need**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0QXjt69WPCXu"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "\n",
        "import gzip\n",
        "import shutil\n",
        "import urllib\n",
        "\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "import pyrodigal\n",
        "\n",
        "import modelseedpy\n",
        "from modelseedpy.core.msgenome import MSGenome\n",
        "\n",
        "from modelseedpy.core.rast_client import RastClient\n",
        "rast = RastClient()\n",
        "\n",
        "from modelseedpy import MSBuilder\n",
        "\n",
        "from dnngior.gapfill_class import *\n",
        "from dnngior.NN_Predictor import NN\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7wvwQfFPE6L6"
      },
      "source": [
        "### **Gather all the data**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "q0xVIFqLWdE7"
      },
      "source": [
        "[Gut colonization data set](https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.57745/O4OJA2)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "collapsed": true,
        "id": "uxXz61TuVVMd",
        "outputId": "c5fee774-c354-4b20-fa0e-2c9cbda85c1c"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "summary": "{\n  \"name\": \"genomes_gc\",\n  \"rows\": 39,\n  \"fields\": [\n    {\n      \"column\": \"SPECIES_AND_STRAIN_NAME \",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 39,\n        \"samples\": [\n          \"Lactococcus_lactis_CIRM-BIA2553 \",\n          \"Hafnia_alvei_CIRM-BIA2828 \",\n          \"Lactiplantibacillus_plantarum_ATCC14431 \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"TAXONOMY\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 25,\n        \"samples\": [\n          \"k__Bacteria|p__Bacillota|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lentilactobacillus|s__Lentilactobacillus_kefiri\",\n          \"k__Bacteria|p__Bacillota|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Pediococcus|s__Pediococcus_parvulus\",\n          \"k__Bacteria|p__Bacillota|c__Bacilli|o__Bacillales|f__Bacillaceae|g__Bacillus|s__Bacillus_pumilus\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \" ASSEMBLY_ACC_NUMBER\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 39,\n        \"samples\": [\n          \" GCA_964062975 \",\n          \" GCA_964063335 \",\n          \" GCA_964084915 \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"BIOPROJECT_ID \",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \" PRJEB74198 \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \" SAMPLE_ID \",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 39,\n        \"samples\": [\n          \" ERS18598077 \"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \" CHROMOSOME_ACC_NUMBER\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 39,\n        \"samples\": [\n          \" OZ061291-OZ061291\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GENOME_COVERAGE\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 36,\n        \"samples\": [\n          \"253x\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GC_PERCENT\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 5.118336489627526,\n        \"min\": 35.0,\n        \"max\": 58.5,\n        \"num_unique_values\": 24,\n        \"samples\": [\n          49.5\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"NB_REPLICONS\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 2,\n        \"min\": 1,\n        \"max\": 11,\n        \"num_unique_values\": 9,\n        \"samples\": [\n          11\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_CHROMOSOME_bp\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 1060521,\n        \"min\": 1585892,\n        \"max\": 6575559,\n        \"num_unique_values\": 39,\n        \"samples\": [\n          2400577\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_01_bp\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 36,\n        \"samples\": [\n          \"565832 (cluster_002)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_02_bp\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 29,\n        \"samples\": [\n          \"110159 (cluster_003)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_03_bp\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 21,\n        \"samples\": [\n          \"13402  (cluster_006)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_04_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 13,\n        \"samples\": [\n          \"10017 (cluster_005)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_05_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 10,\n        \"samples\": [\n          \"5914 (cluster_006)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_06_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 4,\n        \"samples\": [\n          \"8581 (cluster_027)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_07_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"27430 (cluster_008)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_08_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"7633 (cluster_009)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_09_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"10096 (cluster_012)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"SIZE_PLASMID_10_bp\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"7865 (cluster_010)\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}",
              "type": "dataframe",
              "variable_name": "genomes_gc"
            },
            "text/html": [
              "\n",
              "  <div id=\"df-12889ba1-ec79-4a8f-9203-0448fe8a5bef\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>SPECIES_AND_STRAIN_NAME</th>\n",
              "      <th>TAXONOMY</th>\n",
              "      <th>ASSEMBLY_ACC_NUMBER</th>\n",
              "      <th>BIOPROJECT_ID</th>\n",
              "      <th>SAMPLE_ID</th>\n",
              "      <th>CHROMOSOME_ACC_NUMBER</th>\n",
              "      <th>GENOME_COVERAGE</th>\n",
              "      <th>GC_PERCENT</th>\n",
              "      <th>NB_REPLICONS</th>\n",
              "      <th>SIZE_CHROMOSOME_bp</th>\n",
              "      <th>SIZE_PLASMID_01_bp</th>\n",
              "      <th>SIZE_PLASMID_02_bp</th>\n",
              "      <th>SIZE_PLASMID_03_bp</th>\n",
              "      <th>SIZE_PLASMID_04_bp</th>\n",
              "      <th>SIZE_PLASMID_05_bp</th>\n",
              "      <th>SIZE_PLASMID_06_bp</th>\n",
              "      <th>SIZE_PLASMID_07_bp</th>\n",
              "      <th>SIZE_PLASMID_08_bp</th>\n",
              "      <th>SIZE_PLASMID_09_bp</th>\n",
              "      <th>SIZE_PLASMID_10_bp</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Bacillus_pumilus_CIRM-BIA2784</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Bacilla...</td>\n",
              "      <td>GCA_964063375</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598044</td>\n",
              "      <td>OZ061327-OZ061328</td>\n",
              "      <td>330x</td>\n",
              "      <td>41.5</td>\n",
              "      <td>2</td>\n",
              "      <td>3753338</td>\n",
              "      <td>41278 (cluster_002)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Enterococcus_gilvus_CIRM-BIA2700</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062525</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598045</td>\n",
              "      <td>OZ061221-OZ061222</td>\n",
              "      <td>260x</td>\n",
              "      <td>42.0</td>\n",
              "      <td>2</td>\n",
              "      <td>2825867</td>\n",
              "      <td>889036 (cluster_002)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Lacticaseibacillus_paracasei_CIRM-BIA2373</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065205</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598046</td>\n",
              "      <td>OZ061570-OZ061573</td>\n",
              "      <td>440x</td>\n",
              "      <td>46.5</td>\n",
              "      <td>4</td>\n",
              "      <td>3088163</td>\n",
              "      <td>51021  (cluster_002)</td>\n",
              "      <td>28481  (cluster_003)</td>\n",
              "      <td>13402  (cluster_006)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Lactiplantibacillus_pentosus_CNRZ1547</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063425</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598050</td>\n",
              "      <td>OZ061355-OZ061363</td>\n",
              "      <td>215x</td>\n",
              "      <td>46.0</td>\n",
              "      <td>9</td>\n",
              "      <td>3624812</td>\n",
              "      <td>67590  (cluster_002)</td>\n",
              "      <td>54536  (cluster_003)</td>\n",
              "      <td>58048 (cluster_005)</td>\n",
              "      <td>33433  (cluster_006)</td>\n",
              "      <td>33043 (cluster_009)</td>\n",
              "      <td>22967 (cluster_007)</td>\n",
              "      <td>14246  (cluster_008)</td>\n",
              "      <td>3310 (cluster_017)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Lactiplantibacillus_plantarum_ATCC14431</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964084915</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598049</td>\n",
              "      <td>OZ064365-OZ064366</td>\n",
              "      <td>419x</td>\n",
              "      <td>44.5</td>\n",
              "      <td>2</td>\n",
              "      <td>3207663</td>\n",
              "      <td>9254  (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Lactiplantibacillus_plantarum_CIRM-BIA2443</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065245</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598047</td>\n",
              "      <td>OZ061574-OZ061579</td>\n",
              "      <td>288x</td>\n",
              "      <td>44.5</td>\n",
              "      <td>6</td>\n",
              "      <td>3163810</td>\n",
              "      <td>40752 (cluster_002)</td>\n",
              "      <td>33642  (cluster_003)</td>\n",
              "      <td>24389 (cluster_004)</td>\n",
              "      <td>10848 (cluster_005)</td>\n",
              "      <td>8796 (cluster_006)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>Lactiplantibacillus_plantarum_CIRM-BIA2453</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065445</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598048</td>\n",
              "      <td>OZ061611-OZ061617</td>\n",
              "      <td>770x</td>\n",
              "      <td>44.5</td>\n",
              "      <td>7</td>\n",
              "      <td>3259765</td>\n",
              "      <td>52958 (cluster_007)</td>\n",
              "      <td>39456 (cluster_009)</td>\n",
              "      <td>36215 (cluster_013)</td>\n",
              "      <td>26258 (cluster_017)</td>\n",
              "      <td>16539 (cluster_024)</td>\n",
              "      <td>8581 (cluster_027)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>Latilactobacillus_curvatus_CIRM-BIA2781</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063355</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598051</td>\n",
              "      <td>OZ061323-OZ061326</td>\n",
              "      <td>165x</td>\n",
              "      <td>42.0</td>\n",
              "      <td>4</td>\n",
              "      <td>2062891</td>\n",
              "      <td>43624 (cluster_002)</td>\n",
              "      <td>22323 (cluster_003)</td>\n",
              "      <td>12652 (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>Latilactobacillus_curvatus_J116</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063385</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598052</td>\n",
              "      <td>OZ061332-OZ061335</td>\n",
              "      <td>1000x</td>\n",
              "      <td>42.0</td>\n",
              "      <td>4</td>\n",
              "      <td>1928854</td>\n",
              "      <td>58746 (cluster_002)</td>\n",
              "      <td>12649 (cluster_003)</td>\n",
              "      <td>10908 (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>Latilactobacillus_sakei_CIRM-BIA1912</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063755</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598053</td>\n",
              "      <td>OZ061411-OZ061415</td>\n",
              "      <td>1200x</td>\n",
              "      <td>41.5</td>\n",
              "      <td>5</td>\n",
              "      <td>1907497</td>\n",
              "      <td>31814 (cluster_002)</td>\n",
              "      <td>11126 (cluster_003)</td>\n",
              "      <td>12662 (cluster_004)</td>\n",
              "      <td>12470 (cluster_005)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>Latilactobacillus_sakei_J33</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063415</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598054</td>\n",
              "      <td>OZ061338-OZ061340</td>\n",
              "      <td>310x</td>\n",
              "      <td>41.0</td>\n",
              "      <td>3</td>\n",
              "      <td>2036585</td>\n",
              "      <td>44592 (cluster_002)</td>\n",
              "      <td>9151 (cluster_003)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>Lentilactobacillus_buchneri_CIRM-BIA1514</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964060345</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598055</td>\n",
              "      <td>OZ060996-OZ060998</td>\n",
              "      <td>1350x</td>\n",
              "      <td>44.0</td>\n",
              "      <td>3</td>\n",
              "      <td>2501169</td>\n",
              "      <td>79367 (cluster_002)</td>\n",
              "      <td>44732 (cluster_003)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>Lentilactobacillus_buchneri_CIRM-BIA2511</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065095</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598056</td>\n",
              "      <td>OZ061537-OZ061538</td>\n",
              "      <td>340x</td>\n",
              "      <td>44.0</td>\n",
              "      <td>2</td>\n",
              "      <td>2507470</td>\n",
              "      <td>56670 (cluster_002)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>Lentilactobacillus_kefiri_CIRM-BIA2539</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964068245</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598057</td>\n",
              "      <td>OZ061888-OZ061893</td>\n",
              "      <td>290x</td>\n",
              "      <td>42.0</td>\n",
              "      <td>6</td>\n",
              "      <td>2356324</td>\n",
              "      <td>82354 (cluster_002)</td>\n",
              "      <td>42757 (cluster_003)</td>\n",
              "      <td>11006 (cluster_004)</td>\n",
              "      <td>6952 (cluster_005)</td>\n",
              "      <td>4608 (cluster_006)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>Lentilactobacillus_kefiri_CIRM-BIA2556</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062605</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598058</td>\n",
              "      <td>OZ061239-OZ061244</td>\n",
              "      <td>208x</td>\n",
              "      <td>42.0</td>\n",
              "      <td>6</td>\n",
              "      <td>2353226</td>\n",
              "      <td>55933 (cluster_002)</td>\n",
              "      <td>67679 (cluster_003)</td>\n",
              "      <td>37864 (cluster_004)</td>\n",
              "      <td>20585 (cluster_006)</td>\n",
              "      <td>7785 (cluster_007)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>Lentilactobacillus_parabuchneri_CIRM-BIA2637</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062735</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598059</td>\n",
              "      <td>OZ061254-OZ061256</td>\n",
              "      <td>450x</td>\n",
              "      <td>43.5</td>\n",
              "      <td>3</td>\n",
              "      <td>2582132</td>\n",
              "      <td>40963 (cluster_002)</td>\n",
              "      <td>3899 (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>Levilactobacillus_brevis_CIRM-BIA2444</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065285</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598060</td>\n",
              "      <td>OZ061585-OZ061590</td>\n",
              "      <td>295x</td>\n",
              "      <td>46.0</td>\n",
              "      <td>6</td>\n",
              "      <td>2341171</td>\n",
              "      <td>36450 (cluster_002)</td>\n",
              "      <td>35387 (cluster_003)</td>\n",
              "      <td>33150 (cluster_004)</td>\n",
              "      <td>27037 (cluster_005)</td>\n",
              "      <td>3225 (cluster_007)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>Levilactobacillus_brevis_CIRM-BIA2515</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065115</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598061</td>\n",
              "      <td>OZ061544-OZ061547</td>\n",
              "      <td>590x</td>\n",
              "      <td>46.0</td>\n",
              "      <td>4</td>\n",
              "      <td>2388111</td>\n",
              "      <td>55140 (cluster_006)</td>\n",
              "      <td>37854 (cluster_014)</td>\n",
              "      <td>10221 (cluster_031)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>Levilactobacillus_cerevisiae_CIRM-BIA2671</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062745</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598062</td>\n",
              "      <td>OZ061257-OZ061267</td>\n",
              "      <td>510x</td>\n",
              "      <td>49.5</td>\n",
              "      <td>11</td>\n",
              "      <td>3165628</td>\n",
              "      <td>59058 (cluster_002)</td>\n",
              "      <td>42994 (cluster_003)</td>\n",
              "      <td>47169 (cluster_004)</td>\n",
              "      <td>40800 (cluster_005)</td>\n",
              "      <td>40382 (cluster_006)</td>\n",
              "      <td>35722 (cluster_007)</td>\n",
              "      <td>27430 (cluster_008)</td>\n",
              "      <td>7633 (cluster_009)</td>\n",
              "      <td>10096 (cluster_012)</td>\n",
              "      <td>7865 (cluster_010)</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>Levilactobacillus_yonginensis_CIRM-BIA2544</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065165</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598063</td>\n",
              "      <td>OZ061549-OZ061553</td>\n",
              "      <td>710x</td>\n",
              "      <td>47.0</td>\n",
              "      <td>5</td>\n",
              "      <td>2648638</td>\n",
              "      <td>92110 (cluster_002)</td>\n",
              "      <td>43285 (cluster_003)</td>\n",
              "      <td>40534 (cluster_004)</td>\n",
              "      <td>54373 (cluster_005)</td>\n",
              "      <td>6167 (cluster_007)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>Liquorilactobacillus_satsumensis_CIRM-BIA2521</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065465</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598064</td>\n",
              "      <td>OZ061606-OZ061610</td>\n",
              "      <td>250x</td>\n",
              "      <td>40.0</td>\n",
              "      <td>5</td>\n",
              "      <td>2559744</td>\n",
              "      <td>55909 (cluster_002)</td>\n",
              "      <td>23264 (cluster_003)</td>\n",
              "      <td>28020 (cluster_004)</td>\n",
              "      <td>5547 (cluster_006)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>Loigolactobacillus_coryniformis_CIRM-BIA2685</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062795</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598065</td>\n",
              "      <td>OZ061245-OZ061247</td>\n",
              "      <td>260x</td>\n",
              "      <td>43.0</td>\n",
              "      <td>3</td>\n",
              "      <td>2969537</td>\n",
              "      <td>40202 (cluster_002)</td>\n",
              "      <td>15287 (cluster_003)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>Pediococcus_ethanolidurans_CIRM-BIA2669</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062635</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598066</td>\n",
              "      <td>OZ061250-OZ061252</td>\n",
              "      <td>335x</td>\n",
              "      <td>37.5</td>\n",
              "      <td>3</td>\n",
              "      <td>2164028</td>\n",
              "      <td>55735 (cluster_002)</td>\n",
              "      <td>8503 (cluster_005)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>Pediococcus_parvulus_CIRM-BIA2501</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065435</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598067</td>\n",
              "      <td>OZ061592-OZ061593</td>\n",
              "      <td>255x</td>\n",
              "      <td>38.5</td>\n",
              "      <td>2</td>\n",
              "      <td>1938320</td>\n",
              "      <td>38693 (cluster_002)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>Pediococcus_parvulus_F2-sp2-RGJ</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063395</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598068</td>\n",
              "      <td>OZ061336-OZ061337</td>\n",
              "      <td>400x</td>\n",
              "      <td>39.0</td>\n",
              "      <td>2</td>\n",
              "      <td>1875862</td>\n",
              "      <td>30772 (cluster_002)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>Pediococcus_parvulus_IOEB-8801</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063445</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598069</td>\n",
              "      <td>OZ061351-OZ061354</td>\n",
              "      <td>272x</td>\n",
              "      <td>38.5</td>\n",
              "      <td>4</td>\n",
              "      <td>2090870</td>\n",
              "      <td>44543 (cluster_002)</td>\n",
              "      <td>51819 (cluster_003)</td>\n",
              "      <td>27848 (cluster_005)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>Pediococcus_parvulus_IOEB-9646</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063155</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598070</td>\n",
              "      <td>OZ061292-OZ061294</td>\n",
              "      <td>399x</td>\n",
              "      <td>38.5</td>\n",
              "      <td>3</td>\n",
              "      <td>2090911</td>\n",
              "      <td>44550 (cluster_002)</td>\n",
              "      <td>27848 (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>Schleiferilactobacillus_harbinensis_CIRM-BIA2388</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964069135</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598071</td>\n",
              "      <td>OZ061976-OZ061976</td>\n",
              "      <td>450x</td>\n",
              "      <td>53.5</td>\n",
              "      <td>1</td>\n",
              "      <td>3224482</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>Leuconostoc_carnosum_A1b-VegA</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964084875</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598072</td>\n",
              "      <td>OZ064360-OZ064363</td>\n",
              "      <td>257x</td>\n",
              "      <td>37.0</td>\n",
              "      <td>4</td>\n",
              "      <td>1676057</td>\n",
              "      <td>40938 (cluster_002)</td>\n",
              "      <td>28477 (cluster_003)</td>\n",
              "      <td>30820 (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>Leuconostoc_carnosum_CIRM-BIA2809</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063315</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598073</td>\n",
              "      <td>OZ061310-OZ061314</td>\n",
              "      <td>2110x</td>\n",
              "      <td>37.5</td>\n",
              "      <td>5</td>\n",
              "      <td>1585892</td>\n",
              "      <td>55481 (cluster_002)</td>\n",
              "      <td>53754 (cluster_003)</td>\n",
              "      <td>47149 (cluster_004)</td>\n",
              "      <td>25048 (cluster_005)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>Leuconostoc_mesenteroides_FLEC04</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063405</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598074</td>\n",
              "      <td>OZ061345-OZ061350</td>\n",
              "      <td>263x</td>\n",
              "      <td>38.0</td>\n",
              "      <td>6</td>\n",
              "      <td>1848396</td>\n",
              "      <td>40264 (cluster_002)</td>\n",
              "      <td>30801 (cluster_003)</td>\n",
              "      <td>17772 (cluster_004)</td>\n",
              "      <td>10017 (cluster_005)</td>\n",
              "      <td>5914 (cluster_006)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>Leuconostoc_mesenteroides_FME302</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964063435</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598075</td>\n",
              "      <td>OZ061341-OZ061344</td>\n",
              "      <td>447x</td>\n",
              "      <td>38.0</td>\n",
              "      <td>4</td>\n",
              "      <td>1918139</td>\n",
              "      <td>15874 (cluster_003)</td>\n",
              "      <td>15004 (cluster_004)</td>\n",
              "      <td>17857 (cluster_006)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>Leuconostoc_pseudomesenteroides_CIRM-BIA2446</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964065235</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598076</td>\n",
              "      <td>OZ061555-OZ061556</td>\n",
              "      <td>720x</td>\n",
              "      <td>39.0</td>\n",
              "      <td>2</td>\n",
              "      <td>2186513</td>\n",
              "      <td>29175 (cluster_002)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>Lactococcus_lactis_CIRM-BIA2553</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062975</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598077</td>\n",
              "      <td>OZ061291-OZ061291</td>\n",
              "      <td>300x</td>\n",
              "      <td>35.0</td>\n",
              "      <td>1</td>\n",
              "      <td>2400577</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>Lactococcus_lactis_CIRM-BIA2698</td>\n",
              "      <td>k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...</td>\n",
              "      <td>GCA_964062575</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598078</td>\n",
              "      <td>OZ061229-OZ061235</td>\n",
              "      <td>260x</td>\n",
              "      <td>35.5</td>\n",
              "      <td>7</td>\n",
              "      <td>2527645</td>\n",
              "      <td>68742 (cluster_002)</td>\n",
              "      <td>66408 (cluster_003)</td>\n",
              "      <td>61802 (cluster_004)</td>\n",
              "      <td>42478 (cluster_005)</td>\n",
              "      <td>24469 (cluster_006)</td>\n",
              "      <td>7936 (cluster_009)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>Hafnia_alvei_CIRM-BIA2718</td>\n",
              "      <td>k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...</td>\n",
              "      <td>GCA_964063325</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598079</td>\n",
              "      <td>OZ061315-OZ061317</td>\n",
              "      <td>350x</td>\n",
              "      <td>48.5</td>\n",
              "      <td>3</td>\n",
              "      <td>4873592</td>\n",
              "      <td>109260 (cluster_002)</td>\n",
              "      <td>87436 (cluster_003)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>Hafnia_alvei_CIRM-BIA2828</td>\n",
              "      <td>k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...</td>\n",
              "      <td>GCA_964063335</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598080</td>\n",
              "      <td>OZ061319-OZ061322</td>\n",
              "      <td>805x</td>\n",
              "      <td>49.0</td>\n",
              "      <td>4</td>\n",
              "      <td>4725319</td>\n",
              "      <td>133807 (cluster_002)</td>\n",
              "      <td>110159 (cluster_003)</td>\n",
              "      <td>89193 (cluster_004)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>Rahnella_aquatilis_CIRM-BIA2795</td>\n",
              "      <td>k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...</td>\n",
              "      <td>GCA_964063365</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598081</td>\n",
              "      <td>OZ061329-OZ061331</td>\n",
              "      <td>160x</td>\n",
              "      <td>52.5</td>\n",
              "      <td>3</td>\n",
              "      <td>5285568</td>\n",
              "      <td>565832 (cluster_002)</td>\n",
              "      <td>97048 (cluster_010)</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>Pseudomonas_lini_CIRM-BIA2854</td>\n",
              "      <td>k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...</td>\n",
              "      <td>GCA_964063345</td>\n",
              "      <td>PRJEB74198</td>\n",
              "      <td>ERS18598082</td>\n",
              "      <td>OZ061318-OZ061318</td>\n",
              "      <td>253x</td>\n",
              "      <td>58.5</td>\n",
              "      <td>1</td>\n",
              "      <td>6575559</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-12889ba1-ec79-4a8f-9203-0448fe8a5bef')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-12889ba1-ec79-4a8f-9203-0448fe8a5bef button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-12889ba1-ec79-4a8f-9203-0448fe8a5bef');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-30d1e937-60b9-4c5f-873a-f819aa13a0c4\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-30d1e937-60b9-4c5f-873a-f819aa13a0c4')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-30d1e937-60b9-4c5f-873a-f819aa13a0c4 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_2768ce52-733d-48c7-8849-ed04ddb7c285\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('genomes_gc')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_2768ce52-733d-48c7-8849-ed04ddb7c285 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('genomes_gc');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "                             SPECIES_AND_STRAIN_NAME   \\\n",
              "0                      Bacillus_pumilus_CIRM-BIA2784    \n",
              "1                   Enterococcus_gilvus_CIRM-BIA2700    \n",
              "2          Lacticaseibacillus_paracasei_CIRM-BIA2373    \n",
              "3              Lactiplantibacillus_pentosus_CNRZ1547    \n",
              "4            Lactiplantibacillus_plantarum_ATCC14431    \n",
              "5         Lactiplantibacillus_plantarum_CIRM-BIA2443    \n",
              "6         Lactiplantibacillus_plantarum_CIRM-BIA2453    \n",
              "7            Latilactobacillus_curvatus_CIRM-BIA2781    \n",
              "8                    Latilactobacillus_curvatus_J116    \n",
              "9               Latilactobacillus_sakei_CIRM-BIA1912    \n",
              "10                       Latilactobacillus_sakei_J33    \n",
              "11          Lentilactobacillus_buchneri_CIRM-BIA1514    \n",
              "12          Lentilactobacillus_buchneri_CIRM-BIA2511    \n",
              "13            Lentilactobacillus_kefiri_CIRM-BIA2539    \n",
              "14            Lentilactobacillus_kefiri_CIRM-BIA2556    \n",
              "15      Lentilactobacillus_parabuchneri_CIRM-BIA2637    \n",
              "16             Levilactobacillus_brevis_CIRM-BIA2444    \n",
              "17             Levilactobacillus_brevis_CIRM-BIA2515    \n",
              "18         Levilactobacillus_cerevisiae_CIRM-BIA2671    \n",
              "19        Levilactobacillus_yonginensis_CIRM-BIA2544    \n",
              "20     Liquorilactobacillus_satsumensis_CIRM-BIA2521    \n",
              "21      Loigolactobacillus_coryniformis_CIRM-BIA2685    \n",
              "22           Pediococcus_ethanolidurans_CIRM-BIA2669    \n",
              "23                 Pediococcus_parvulus_CIRM-BIA2501    \n",
              "24                   Pediococcus_parvulus_F2-sp2-RGJ    \n",
              "25                    Pediococcus_parvulus_IOEB-8801    \n",
              "26                    Pediococcus_parvulus_IOEB-9646    \n",
              "27  Schleiferilactobacillus_harbinensis_CIRM-BIA2388    \n",
              "28                     Leuconostoc_carnosum_A1b-VegA    \n",
              "29                 Leuconostoc_carnosum_CIRM-BIA2809    \n",
              "30                  Leuconostoc_mesenteroides_FLEC04    \n",
              "31                  Leuconostoc_mesenteroides_FME302    \n",
              "32      Leuconostoc_pseudomesenteroides_CIRM-BIA2446    \n",
              "33                   Lactococcus_lactis_CIRM-BIA2553    \n",
              "34                   Lactococcus_lactis_CIRM-BIA2698    \n",
              "35                         Hafnia_alvei_CIRM-BIA2718    \n",
              "36                         Hafnia_alvei_CIRM-BIA2828    \n",
              "37                   Rahnella_aquatilis_CIRM-BIA2795    \n",
              "38                     Pseudomonas_lini_CIRM-BIA2854    \n",
              "\n",
              "                                             TAXONOMY  ASSEMBLY_ACC_NUMBER  \\\n",
              "0   k__Bacteria|p__Bacillota|c__Bacilli|o__Bacilla...       GCA_964063375    \n",
              "1   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062525    \n",
              "2   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065205    \n",
              "3   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063425    \n",
              "4   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964084915    \n",
              "5   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065245    \n",
              "6   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065445    \n",
              "7   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063355    \n",
              "8   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063385    \n",
              "9   k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063755    \n",
              "10  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063415    \n",
              "11  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964060345    \n",
              "12  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065095    \n",
              "13  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964068245    \n",
              "14  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062605    \n",
              "15  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062735    \n",
              "16  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065285    \n",
              "17  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065115    \n",
              "18  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062745    \n",
              "19  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065165    \n",
              "20  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065465    \n",
              "21  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062795    \n",
              "22  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062635    \n",
              "23  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065435    \n",
              "24  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063395    \n",
              "25  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063445    \n",
              "26  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063155    \n",
              "27  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964069135    \n",
              "28  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964084875    \n",
              "29  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063315    \n",
              "30  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063405    \n",
              "31  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964063435    \n",
              "32  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964065235    \n",
              "33  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062975    \n",
              "34  k__Bacteria|p__Bacillota|c__Bacilli|o__Lactoba...       GCA_964062575    \n",
              "35  k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...       GCA_964063325    \n",
              "36  k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...       GCA_964063335    \n",
              "37  k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...       GCA_964063365    \n",
              "38  k__Bacteria|p__Pseudomonadota|c__Gammaproteoba...       GCA_964063345    \n",
              "\n",
              "   BIOPROJECT_ID      SAMPLE_ID   CHROMOSOME_ACC_NUMBER GENOME_COVERAGE  \\\n",
              "0     PRJEB74198    ERS18598044       OZ061327-OZ061328            330x   \n",
              "1     PRJEB74198    ERS18598045       OZ061221-OZ061222            260x   \n",
              "2     PRJEB74198    ERS18598046       OZ061570-OZ061573            440x   \n",
              "3     PRJEB74198    ERS18598050       OZ061355-OZ061363            215x   \n",
              "4     PRJEB74198    ERS18598049       OZ064365-OZ064366            419x   \n",
              "5     PRJEB74198    ERS18598047       OZ061574-OZ061579            288x   \n",
              "6     PRJEB74198    ERS18598048       OZ061611-OZ061617            770x   \n",
              "7     PRJEB74198    ERS18598051       OZ061323-OZ061326            165x   \n",
              "8     PRJEB74198    ERS18598052       OZ061332-OZ061335           1000x   \n",
              "9     PRJEB74198    ERS18598053       OZ061411-OZ061415           1200x   \n",
              "10    PRJEB74198    ERS18598054       OZ061338-OZ061340            310x   \n",
              "11    PRJEB74198    ERS18598055       OZ060996-OZ060998           1350x   \n",
              "12    PRJEB74198    ERS18598056       OZ061537-OZ061538            340x   \n",
              "13    PRJEB74198    ERS18598057       OZ061888-OZ061893            290x   \n",
              "14    PRJEB74198    ERS18598058       OZ061239-OZ061244            208x   \n",
              "15    PRJEB74198    ERS18598059       OZ061254-OZ061256            450x   \n",
              "16    PRJEB74198    ERS18598060       OZ061585-OZ061590            295x   \n",
              "17    PRJEB74198    ERS18598061       OZ061544-OZ061547            590x   \n",
              "18    PRJEB74198    ERS18598062       OZ061257-OZ061267            510x   \n",
              "19    PRJEB74198    ERS18598063       OZ061549-OZ061553            710x   \n",
              "20    PRJEB74198    ERS18598064       OZ061606-OZ061610            250x   \n",
              "21    PRJEB74198    ERS18598065       OZ061245-OZ061247            260x   \n",
              "22    PRJEB74198    ERS18598066       OZ061250-OZ061252            335x   \n",
              "23    PRJEB74198    ERS18598067       OZ061592-OZ061593            255x   \n",
              "24    PRJEB74198    ERS18598068       OZ061336-OZ061337            400x   \n",
              "25    PRJEB74198    ERS18598069       OZ061351-OZ061354            272x   \n",
              "26    PRJEB74198    ERS18598070       OZ061292-OZ061294            399x   \n",
              "27    PRJEB74198    ERS18598071       OZ061976-OZ061976            450x   \n",
              "28    PRJEB74198    ERS18598072       OZ064360-OZ064363            257x   \n",
              "29    PRJEB74198    ERS18598073       OZ061310-OZ061314           2110x   \n",
              "30    PRJEB74198    ERS18598074       OZ061345-OZ061350            263x   \n",
              "31    PRJEB74198    ERS18598075       OZ061341-OZ061344            447x   \n",
              "32    PRJEB74198    ERS18598076       OZ061555-OZ061556            720x   \n",
              "33    PRJEB74198    ERS18598077       OZ061291-OZ061291            300x   \n",
              "34    PRJEB74198    ERS18598078       OZ061229-OZ061235            260x   \n",
              "35    PRJEB74198    ERS18598079       OZ061315-OZ061317            350x   \n",
              "36    PRJEB74198    ERS18598080       OZ061319-OZ061322            805x   \n",
              "37    PRJEB74198    ERS18598081       OZ061329-OZ061331            160x   \n",
              "38    PRJEB74198    ERS18598082       OZ061318-OZ061318            253x   \n",
              "\n",
              "    GC_PERCENT  NB_REPLICONS  SIZE_CHROMOSOME_bp    SIZE_PLASMID_01_bp  \\\n",
              "0         41.5             2             3753338   41278 (cluster_002)   \n",
              "1         42.0             2             2825867  889036 (cluster_002)   \n",
              "2         46.5             4             3088163  51021  (cluster_002)   \n",
              "3         46.0             9             3624812  67590  (cluster_002)   \n",
              "4         44.5             2             3207663   9254  (cluster_004)   \n",
              "5         44.5             6             3163810   40752 (cluster_002)   \n",
              "6         44.5             7             3259765   52958 (cluster_007)   \n",
              "7         42.0             4             2062891   43624 (cluster_002)   \n",
              "8         42.0             4             1928854   58746 (cluster_002)   \n",
              "9         41.5             5             1907497   31814 (cluster_002)   \n",
              "10        41.0             3             2036585   44592 (cluster_002)   \n",
              "11        44.0             3             2501169   79367 (cluster_002)   \n",
              "12        44.0             2             2507470   56670 (cluster_002)   \n",
              "13        42.0             6             2356324   82354 (cluster_002)   \n",
              "14        42.0             6             2353226   55933 (cluster_002)   \n",
              "15        43.5             3             2582132   40963 (cluster_002)   \n",
              "16        46.0             6             2341171   36450 (cluster_002)   \n",
              "17        46.0             4             2388111   55140 (cluster_006)   \n",
              "18        49.5            11             3165628   59058 (cluster_002)   \n",
              "19        47.0             5             2648638   92110 (cluster_002)   \n",
              "20        40.0             5             2559744   55909 (cluster_002)   \n",
              "21        43.0             3             2969537   40202 (cluster_002)   \n",
              "22        37.5             3             2164028   55735 (cluster_002)   \n",
              "23        38.5             2             1938320   38693 (cluster_002)   \n",
              "24        39.0             2             1875862   30772 (cluster_002)   \n",
              "25        38.5             4             2090870   44543 (cluster_002)   \n",
              "26        38.5             3             2090911   44550 (cluster_002)   \n",
              "27        53.5             1             3224482                   NaN   \n",
              "28        37.0             4             1676057   40938 (cluster_002)   \n",
              "29        37.5             5             1585892   55481 (cluster_002)   \n",
              "30        38.0             6             1848396   40264 (cluster_002)   \n",
              "31        38.0             4             1918139   15874 (cluster_003)   \n",
              "32        39.0             2             2186513   29175 (cluster_002)   \n",
              "33        35.0             1             2400577                   NaN   \n",
              "34        35.5             7             2527645   68742 (cluster_002)   \n",
              "35        48.5             3             4873592  109260 (cluster_002)   \n",
              "36        49.0             4             4725319  133807 (cluster_002)   \n",
              "37        52.5             3             5285568  565832 (cluster_002)   \n",
              "38        58.5             1             6575559                   NaN   \n",
              "\n",
              "      SIZE_PLASMID_02_bp    SIZE_PLASMID_03_bp    SIZE_PLASMID_04_bp  \\\n",
              "0                    NaN                   NaN                   NaN   \n",
              "1                    NaN                   NaN                   NaN   \n",
              "2   28481  (cluster_003)  13402  (cluster_006)                   NaN   \n",
              "3   54536  (cluster_003)   58048 (cluster_005)  33433  (cluster_006)   \n",
              "4                    NaN                   NaN                   NaN   \n",
              "5   33642  (cluster_003)   24389 (cluster_004)   10848 (cluster_005)   \n",
              "6    39456 (cluster_009)   36215 (cluster_013)   26258 (cluster_017)   \n",
              "7    22323 (cluster_003)   12652 (cluster_004)                   NaN   \n",
              "8    12649 (cluster_003)   10908 (cluster_004)                   NaN   \n",
              "9    11126 (cluster_003)   12662 (cluster_004)   12470 (cluster_005)   \n",
              "10    9151 (cluster_003)                   NaN                   NaN   \n",
              "11   44732 (cluster_003)                   NaN                   NaN   \n",
              "12                   NaN                   NaN                   NaN   \n",
              "13   42757 (cluster_003)   11006 (cluster_004)    6952 (cluster_005)   \n",
              "14   67679 (cluster_003)   37864 (cluster_004)   20585 (cluster_006)   \n",
              "15    3899 (cluster_004)                   NaN                   NaN   \n",
              "16   35387 (cluster_003)   33150 (cluster_004)   27037 (cluster_005)   \n",
              "17   37854 (cluster_014)   10221 (cluster_031)                   NaN   \n",
              "18   42994 (cluster_003)   47169 (cluster_004)   40800 (cluster_005)   \n",
              "19   43285 (cluster_003)   40534 (cluster_004)   54373 (cluster_005)   \n",
              "20   23264 (cluster_003)   28020 (cluster_004)    5547 (cluster_006)   \n",
              "21   15287 (cluster_003)                   NaN                   NaN   \n",
              "22    8503 (cluster_005)                   NaN                   NaN   \n",
              "23                   NaN                   NaN                   NaN   \n",
              "24                   NaN                   NaN                   NaN   \n",
              "25   51819 (cluster_003)   27848 (cluster_005)                   NaN   \n",
              "26   27848 (cluster_004)                   NaN                   NaN   \n",
              "27                   NaN                   NaN                   NaN   \n",
              "28   28477 (cluster_003)   30820 (cluster_004)                   NaN   \n",
              "29   53754 (cluster_003)   47149 (cluster_004)   25048 (cluster_005)   \n",
              "30   30801 (cluster_003)   17772 (cluster_004)   10017 (cluster_005)   \n",
              "31   15004 (cluster_004)   17857 (cluster_006)                   NaN   \n",
              "32                   NaN                   NaN                   NaN   \n",
              "33                   NaN                   NaN                   NaN   \n",
              "34   66408 (cluster_003)   61802 (cluster_004)   42478 (cluster_005)   \n",
              "35   87436 (cluster_003)                   NaN                   NaN   \n",
              "36  110159 (cluster_003)   89193 (cluster_004)                   NaN   \n",
              "37   97048 (cluster_010)                   NaN                   NaN   \n",
              "38                   NaN                   NaN                   NaN   \n",
              "\n",
              "     SIZE_PLASMID_05_bp   SIZE_PLASMID_06_bp    SIZE_PLASMID_07_bp  \\\n",
              "0                   NaN                  NaN                   NaN   \n",
              "1                   NaN                  NaN                   NaN   \n",
              "2                   NaN                  NaN                   NaN   \n",
              "3   33043 (cluster_009)  22967 (cluster_007)  14246  (cluster_008)   \n",
              "4                   NaN                  NaN                   NaN   \n",
              "5    8796 (cluster_006)                  NaN                   NaN   \n",
              "6   16539 (cluster_024)   8581 (cluster_027)                   NaN   \n",
              "7                   NaN                  NaN                   NaN   \n",
              "8                   NaN                  NaN                   NaN   \n",
              "9                   NaN                  NaN                   NaN   \n",
              "10                  NaN                  NaN                   NaN   \n",
              "11                  NaN                  NaN                   NaN   \n",
              "12                  NaN                  NaN                   NaN   \n",
              "13   4608 (cluster_006)                  NaN                   NaN   \n",
              "14   7785 (cluster_007)                  NaN                   NaN   \n",
              "15                  NaN                  NaN                   NaN   \n",
              "16   3225 (cluster_007)                  NaN                   NaN   \n",
              "17                  NaN                  NaN                   NaN   \n",
              "18  40382 (cluster_006)  35722 (cluster_007)   27430 (cluster_008)   \n",
              "19   6167 (cluster_007)                  NaN                   NaN   \n",
              "20                  NaN                  NaN                   NaN   \n",
              "21                  NaN                  NaN                   NaN   \n",
              "22                  NaN                  NaN                   NaN   \n",
              "23                  NaN                  NaN                   NaN   \n",
              "24                  NaN                  NaN                   NaN   \n",
              "25                  NaN                  NaN                   NaN   \n",
              "26                  NaN                  NaN                   NaN   \n",
              "27                  NaN                  NaN                   NaN   \n",
              "28                  NaN                  NaN                   NaN   \n",
              "29                  NaN                  NaN                   NaN   \n",
              "30   5914 (cluster_006)                  NaN                   NaN   \n",
              "31                  NaN                  NaN                   NaN   \n",
              "32                  NaN                  NaN                   NaN   \n",
              "33                  NaN                  NaN                   NaN   \n",
              "34  24469 (cluster_006)   7936 (cluster_009)                   NaN   \n",
              "35                  NaN                  NaN                   NaN   \n",
              "36                  NaN                  NaN                   NaN   \n",
              "37                  NaN                  NaN                   NaN   \n",
              "38                  NaN                  NaN                   NaN   \n",
              "\n",
              "    SIZE_PLASMID_08_bp   SIZE_PLASMID_09_bp  SIZE_PLASMID_10_bp  \n",
              "0                  NaN                  NaN                 NaN  \n",
              "1                  NaN                  NaN                 NaN  \n",
              "2                  NaN                  NaN                 NaN  \n",
              "3   3310 (cluster_017)                  NaN                 NaN  \n",
              "4                  NaN                  NaN                 NaN  \n",
              "5                  NaN                  NaN                 NaN  \n",
              "6                  NaN                  NaN                 NaN  \n",
              "7                  NaN                  NaN                 NaN  \n",
              "8                  NaN                  NaN                 NaN  \n",
              "9                  NaN                  NaN                 NaN  \n",
              "10                 NaN                  NaN                 NaN  \n",
              "11                 NaN                  NaN                 NaN  \n",
              "12                 NaN                  NaN                 NaN  \n",
              "13                 NaN                  NaN                 NaN  \n",
              "14                 NaN                  NaN                 NaN  \n",
              "15                 NaN                  NaN                 NaN  \n",
              "16                 NaN                  NaN                 NaN  \n",
              "17                 NaN                  NaN                 NaN  \n",
              "18  7633 (cluster_009)  10096 (cluster_012)  7865 (cluster_010)  \n",
              "19                 NaN                  NaN                 NaN  \n",
              "20                 NaN                  NaN                 NaN  \n",
              "21                 NaN                  NaN                 NaN  \n",
              "22                 NaN                  NaN                 NaN  \n",
              "23                 NaN                  NaN                 NaN  \n",
              "24                 NaN                  NaN                 NaN  \n",
              "25                 NaN                  NaN                 NaN  \n",
              "26                 NaN                  NaN                 NaN  \n",
              "27                 NaN                  NaN                 NaN  \n",
              "28                 NaN                  NaN                 NaN  \n",
              "29                 NaN                  NaN                 NaN  \n",
              "30                 NaN                  NaN                 NaN  \n",
              "31                 NaN                  NaN                 NaN  \n",
              "32                 NaN                  NaN                 NaN  \n",
              "33                 NaN                  NaN                 NaN  \n",
              "34                 NaN                  NaN                 NaN  \n",
              "35                 NaN                  NaN                 NaN  \n",
              "36                 NaN                  NaN                 NaN  \n",
              "37                 NaN                  NaN                 NaN  \n",
              "38                 NaN                  NaN                 NaN  "
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "genomes_gc = pd.read_csv(\"metabolic_toy_model/files/metasimfood_genomic_data.csv\", sep = ';')\n",
        "genomes_gc"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jIus_YCMJI1G"
      },
      "source": [
        "Methanol dataset\n",
        "\n",
        "[see modelSEED biochemistry entry for methanol](https://modelseed.org/biochem/compounds/cpd00116)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "rBzD0bFRJZPe",
        "outputId": "4a367fb5-d404-4d9f-d611-b48546b3c33b"
      },
      "outputs": [
        {
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "genomes_mt"
            },
            "text/html": [
              "\n",
              "  <div id=\"df-a09ae9f0-98a0-4b46-a117-d865b4a3ce95\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>C</th>\n",
              "      <th>k127_911026</th>\n",
              "      <th>1797677</th>\n",
              "      <th>Bacteria; Firmicutes; Clostridiales; Clostridia; NA; NA; Clostridiales bacterium GWB2_37_7;</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>C</td>\n",
              "      <td>k127_121496</td>\n",
              "      <td>1036679</td>\n",
              "      <td>Archaea; Euryarchaeota; Methanosarcinales; Met...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>C</td>\n",
              "      <td>k127_1154015</td>\n",
              "      <td>1036679</td>\n",
              "      <td>Archaea; Euryarchaeota; Methanosarcinales; Met...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>C</td>\n",
              "      <td>k127_911050</td>\n",
              "      <td>1892845</td>\n",
              "      <td>Bacteria; Proteobacteria; Rhizobiales; Alphapr...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>C</td>\n",
              "      <td>k127_546681</td>\n",
              "      <td>51616</td>\n",
              "      <td>Bacteria; Firmicutes; Clostridiales; Clostridi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>C</td>\n",
              "      <td>k127_125</td>\n",
              "      <td>53463</td>\n",
              "      <td>Bacteria; Proteobacteria; Rhodobacterales; Alp...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a09ae9f0-98a0-4b46-a117-d865b4a3ce95')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-a09ae9f0-98a0-4b46-a117-d865b4a3ce95 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-a09ae9f0-98a0-4b46-a117-d865b4a3ce95');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-5843ea7f-5e58-434e-83a9-fb8abb7b2444\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-5843ea7f-5e58-434e-83a9-fb8abb7b2444')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-5843ea7f-5e58-434e-83a9-fb8abb7b2444 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "text/plain": [
              "   C   k127_911026  1797677  \\\n",
              "0  C   k127_121496  1036679   \n",
              "1  C  k127_1154015  1036679   \n",
              "2  C   k127_911050  1892845   \n",
              "3  C   k127_546681    51616   \n",
              "4  C      k127_125    53463   \n",
              "\n",
              "  Bacteria; Firmicutes; Clostridiales; Clostridia; NA; NA; Clostridiales bacterium GWB2_37_7;   \n",
              "0  Archaea; Euryarchaeota; Methanosarcinales; Met...                                            \n",
              "1  Archaea; Euryarchaeota; Methanosarcinales; Met...                                            \n",
              "2  Bacteria; Proteobacteria; Rhizobiales; Alphapr...                                            \n",
              "3  Bacteria; Firmicutes; Clostridiales; Clostridi...                                            \n",
              "4  Bacteria; Proteobacteria; Rhodobacterales; Alp...                                            "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "genomes_mt = pd.read_csv(\"metabolic_toy_model/files/contigs_coassembly-taxNames.tsv\", sep = '\\t')\n",
        "genomes_mt.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Nieu_4WKLzM5"
      },
      "source": [
        "### **Write a function to download the genomes**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZTA0n-MrMH-8"
      },
      "source": [
        "Three bins from the methanol project are in the folder `metabolic_toy_model/files/bins` (only to as examples)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "yorQQxwJXGMQ"
      },
      "outputs": [],
      "source": [
        "def download_fasta(url, gz_id, output):\n",
        "  urllib.request.urlretrieve(url, gz_id)\n",
        "  with gzip.open(gz_id, 'rb') as f_in:\n",
        "      with open(output, 'wb') as f_out:\n",
        "          shutil.copyfileobj(f_in, f_out)\n",
        "  os.remove(gz_id)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qhr82B20NHTl"
      },
      "source": [
        "### **Get the *Bifidobacterium* and the three first genomes from the list above in the a folder called food_bac_genomes**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NxjQGxl2NGmU",
        "outputId": "9657966f-830d-4256-f413-e44e766675ea"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "total 4468\n",
            "drwxr-xr-x 2 root root    4096 Mar 13 08:25 .\n",
            "drwxr-xr-x 1 root root    4096 Mar 13 08:25 ..\n",
            "-rw-r--r-- 1 root root  736519 Mar 13 08:25 GCF_000010425.1_ASM1042v1_protein.faa\n",
            "-rw-r--r-- 1 root root 1342664 Mar 13 08:25 GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700_protein.faa\n",
            "-rw-r--r-- 1 root root 1385571 Mar 13 08:25 GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784_protein.faa\n",
            "-rw-r--r-- 1 root root 1095910 Mar 13 08:25 GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373_protein.faa\n"
          ]
        }
      ],
      "source": [
        "!mkdir food_bac_genomes\n",
        "\n",
        "download_fasta(\"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/010/425/GCF_000010425.1_ASM1042v1/GCF_000010425.1_ASM1042v1_protein.faa.gz\",\n",
        "               \"GCF_000010425.1_ASM1042v1_protein.faa.gz\",\n",
        "               \"food_bac_genomes/GCF_000010425.1_ASM1042v1_protein.faa\")\n",
        "\n",
        "download_fasta(\"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/964/063/375/GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784/GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784_protein.faa.gz\",\n",
        "               \"GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784_protein.faa.gz\",\n",
        "               \"food_bac_genomes/GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784_protein.faa\")\n",
        "\n",
        "download_fasta(\"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/964/062/525/GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700/GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700_protein.faa.gz\",\n",
        "               \"GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700_protein.faa.gz\",\n",
        "               \"food_bac_genomes/GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700_protein.faa\")\n",
        "\n",
        "\n",
        "download_fasta(\"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/964/065/205/GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373/GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373_protein.faa.gz\",\n",
        "               \"GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373_protein.faa.gz\",\n",
        "               \"food_bac_genomes/GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373_protein.faa\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "!ls food_bac_genomes -la"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yuuW-RY6OgrX"
      },
      "source": [
        "#### **Write a function to reconstruct draft models**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Zom-aDT0FXdc"
      },
      "source": [
        "We can make a small modification to the `reconstruct_draft_model` function we used yesterday to give the condition `is_dna_fasta` to know if we should annotate ORFs."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "collapsed": true,
        "id": "OilM6a8oCYDC"
      },
      "outputs": [],
      "source": [
        "def reconstruct_draft_model(model_id, input_fasta_file, output_model_sbml, is_dna_fasta = False):\n",
        "\n",
        "  if is_dna_fasta:\n",
        "    # Parse all records from the FASTA file\n",
        "    records = list(SeqIO.parse(input_fasta_file, \"fasta\"))\n",
        "    # Concatenate all sequence strings\n",
        "    concatenated_sequence = \"\".join(str(record.seq) for record in records)\n",
        "    concatenated_record = SeqRecord(Seq(concatenated_sequence),\n",
        "                                id=\"concatenated\",\n",
        "                                description=\"Concatenated sequence of all records\")\n",
        "    dna_seq = str(concatenated_record.seq)  # Convert the sequence to a plain string\n",
        "    gene_finder = pyrodigal.GeneFinder()\n",
        "    gene_finder.train(dna_seq)\n",
        "    genes = gene_finder.find_genes(dna_seq)\n",
        "\n",
        "    with open(input_fasta_file + \".pep.fa\", \"w\") as f:\n",
        "      genes.write_translations(f, sequence_id=\"seqXYZ\")\n",
        "\n",
        "    input_fasta_file = input_fasta_file + \".pep.fa\"\n",
        "\n",
        "  genome = MSGenome.from_fasta(input_fasta_file, split = ' ')\n",
        "  rast.annotate_genome(genome)\n",
        "\n",
        "  base_model = MSBuilder.build_metabolic_model(model_id = model_id,\n",
        "                                             genome   = genome,\n",
        "                                             index    = \"0\",\n",
        "                                             classic_biomass = True,\n",
        "                                             gapfill_model   = False,\n",
        "                                             gapfill_media   = None,\n",
        "                                             annotate_with_rast = True,\n",
        "                                             allow_all_non_grp_reactions = True\n",
        "                                            )\n",
        "\n",
        "  cobra.io.write_sbml_model(cobra_model = base_model, filename = output_model_sbml)\n",
        "\n",
        "  return base_model"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GDobbLp2P8zQ"
      },
      "source": [
        "### **Build draft and gapfilled models**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PBcSuKT3QC7T"
      },
      "source": [
        "##### **gut colonization**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lZ-is4saQHc_"
      },
      "source": [
        "we place them in the folder `food_bac_draft_models` and `food_bac_gapfilled_models`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "-nJW0kXdaQgs",
        "outputId": "f474fd35-2a4a-4c25-ff88-934372e6cc83"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "WARNING:modelseedpy.helpers:mkdir: /usr/local/lib/python3.11/dist-packages/modelseedpy//data/ml\n",
            "WARNING:modelseedpy.helpers:downloading data file to: /usr/local/lib/python3.11/dist-packages/modelseedpy//data/ml/knn_ACNP_RAST_filter_01_17_2023.pickle\n",
            "WARNING:modelseedpy.helpers:downloading data file to: /usr/local/lib/python3.11/dist-packages/modelseedpy//data/ml/knn_ACNP_RAST_filter_01_17_2023_features.json\n",
            "WARNING:modelseedpy.helpers:mkdir: /usr/local/lib/python3.11/dist-packages/modelseedpy//data/templates\n",
            "WARNING:modelseedpy.helpers:downloading data file to: /usr/local/lib/python3.11/dist-packages/modelseedpy//data/templates/template_gram_pos.json\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: _PyDrive2ImportHook.find_spec() not found; falling back to find_module()\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: _PyDriveImportHook.find_spec() not found; falling back to find_module()\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: _GenerativeAIImportHook.find_spec() not found; falling back to find_module()\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: _OpenCVImportHook.find_spec() not found; falling back to find_module()\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: APICoreClientInfoImportHook.find_spec() not found; falling back to find_module()\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: _BokehImportHook.find_spec() not found; falling back to find_module()\n",
            "<frozen importlib._bootstrap>:1047: ImportWarning: _AltairImportHook.find_spec() not found; falling back to find_module()\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  99 / 1007\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  78616 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39308 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19654 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9827 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4913 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2456 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1228 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  614 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  307 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  153 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  76 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.063950.\n",
            "Read LP format model from file /tmp/tmp6r012roo.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 1005 rows, 2144 columns, 9264 nonzeros\n",
            "NN gapfilling added 65 new reactions\n",
            "The NN gapfilled model, comes with 1072 reactions and 1005 metabolites\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  85 / 877\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  78858 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39429 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19714 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9857 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4928 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2464 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1232 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  616 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  308 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  154 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  77 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.051622.\n",
            "Read LP format model from file /tmp/tmpb6o46txk.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 906 rows, 1888 columns, 8198 nonzeros\n",
            "NN gapfilling added 67 new reactions\n",
            "The NN gapfilled model, comes with 944 reactions and 906 metabolites\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  65 / 725\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  79158 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39579 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19789 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9894 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4947 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2473 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1236 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  618 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  309 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  154 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  77 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.061184.\n",
            "Read LP format model from file /tmp/tmpvm2wa4yg.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 786 rows, 1616 columns, 7116 nonzeros\n",
            "NN gapfilling added 83 new reactions\n",
            "The NN gapfilled model, comes with 808 reactions and 786 metabolites\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  98 / 1219\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  78277 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39138 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19569 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9784 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4892 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2446 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1223 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  611 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  305 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  152 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  76 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  38 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.071163.\n",
            "Read LP format model from file /tmp/tmp77qn7b1h.lp\n",
            "Reading time = 0.01 seconds\n",
            ": 1168 rows, 2536 columns, 11322 nonzeros\n",
            "NN gapfilling added 49 new reactions\n",
            "The NN gapfilled model, comes with 1268 reactions and 1168 metabolites\n"
          ]
        }
      ],
      "source": [
        "!mkdir food_bac_draft_models\n",
        "!mkdir food_bac_gapfilled_models\n",
        "\n",
        "\n",
        "genomes = os.listdir(\"food_bac_genomes\")\n",
        "\n",
        "for genome in genomes:\n",
        "  model_id = genome.replace(\"_protein.faa\", \"\")\n",
        "  input_protein_fasta = \"food_bac_genomes/\" + genome\n",
        "  output_model_sbml = \"food_bac_draft_models/\" + model_id + \"_draft.sbml\"\n",
        "  draftModel = reconstruct_draft_model(model_id, input_protein_fasta, output_model_sbml)\n",
        "  gapfilledModel = Gapfill(draftModel = output_model_sbml)\n",
        "  cobra.io.write_sbml_model(gapfilledModel.gapfilledModel, filename = \"food_bac_gapfilled_models/\" + model_id + \"_gapfilled.sbml\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "-qtEU2obVgnc",
        "outputId": "15b3b2f5-a7dc-4086-b3f1-de5f761bfd0b"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "total 9576\n",
            "drwxr-xr-x 2 root root    4096 Mar 13 08:31 .\n",
            "drwxr-xr-x 1 root root    4096 Mar 13 08:25 ..\n",
            "-rw-r--r-- 1 root root 1846499 Mar 13 08:29 GCF_000010425.1_ASM1042v1_draft.sbml\n",
            "-rw-r--r-- 1 root root 2591373 Mar 13 08:25 GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700_draft.sbml\n",
            "-rw-r--r-- 1 root root 3085987 Mar 13 08:31 GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784_draft.sbml\n",
            "-rw-r--r-- 1 root root 2267937 Mar 13 08:27 GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373_draft.sbml\n",
            "total 19268\n",
            "drwxr-xr-x 2 root root    4096 Mar 13 08:32 .\n",
            "drwxr-xr-x 1 root root    4096 Mar 13 08:25 ..\n",
            "-rw-r--r-- 1 root root 4122578 Mar 13 08:30 GCF_000010425.1_ASM1042v1_gapfilled.sbml\n",
            "-rw-r--r-- 1 root root 5141865 Mar 13 08:27 GCF_964062525.1_Enterococcus_gilvus_CIRM-BIA2700_gapfilled.sbml\n",
            "-rw-r--r-- 1 root root 5802216 Mar 13 08:32 GCF_964063375.1_Bacillus_pumilus_CIRM-BIA2784_gapfilled.sbml\n",
            "-rw-r--r-- 1 root root 4646354 Mar 13 08:28 GCF_964065205.1_Lacticaseibacillus_paracasei_CIRM-BIA2373_gapfilled.sbml\n"
          ]
        }
      ],
      "source": [
        "!ls -la food_bac_draft_models\n",
        "!ls -la food_bac_gapfilled_models"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zx1-p776WaDW"
      },
      "source": [
        "##### **Methanol bins**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LKvZkJpmWhQH"
      },
      "source": [
        "We place them in the folder `methanol_draft_models` and `methanol_gapfilled_models`. Set the parameter `is_dna_fasta` to `True`"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "i6sllBsGWsf_",
        "outputId": "84a9cd87-f8a2-4105-cc2a-0626140fb09f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "mkdir: cannot create directory ‘methanol_draft_models’: File exists\n",
            "mkdir: cannot create directory ‘methanol_gapfilled_models’: File exists\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "<ipython-input-22-c9db05c0042a>:5: ResourceWarning: unclosed file <_io.TextIOWrapper name='metabolic_toy_model/files/bins/bin.2.fa' mode='rt' encoding='UTF-8'>\n",
            "  records = list(SeqIO.parse(input_fasta_file, \"fasta\"))\n",
            "ResourceWarning: Enable tracemalloc to get the object allocation traceback\n",
            "WARNING:modelseedpy.helpers:downloading data file to: /usr/local/lib/python3.11/dist-packages/modelseedpy//data/templates/template_gram_neg.json\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  22 / 110\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  80216 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  40108 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  20054 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  10027 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  5013 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2506 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1253 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  626 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  313 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  156 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  78 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  5 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  3 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.250704.\n",
            "Read LP format model from file /tmp/tmpgx7mo66r.lp\n",
            "Reading time = 0.00 seconds\n",
            ": 360 rows, 632 columns, 2876 nonzeros\n",
            "NN gapfilling added 206 new reactions\n",
            "The NN gapfilled model, comes with 316 reactions and 360 metabolites\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "<ipython-input-22-c9db05c0042a>:5: ResourceWarning: unclosed file <_io.TextIOWrapper name='metabolic_toy_model/files/bins/bin.3.fa' mode='rt' encoding='UTF-8'>\n",
            "  records = list(SeqIO.parse(input_fasta_file, \"fasta\"))\n",
            "ResourceWarning: Enable tracemalloc to get the object allocation traceback\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  25 / 282\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  79907 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39953 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19976 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9988 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4994 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2497 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1248 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  624 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  312 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  156 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  78 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  4 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.000000.\n",
            "Read LP format model from file /tmp/tmpw53d2xuh.lp\n",
            "Reading time = 0.00 seconds\n",
            ": 547 rows, 916 columns, 4188 nonzeros\n",
            "NN gapfilling added 176 new reactions\n",
            "The NN gapfilled model, comes with 458 reactions and 547 metabolites\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "<ipython-input-22-c9db05c0042a>:5: ResourceWarning: unclosed file <_io.TextIOWrapper name='metabolic_toy_model/files/bins/bin.1.fa' mode='rt' encoding='UTF-8'>\n",
            "  records = list(SeqIO.parse(input_fasta_file, \"fasta\"))\n",
            "ResourceWarning: Enable tracemalloc to get the object allocation traceback\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n",
            "ERROR:cobra.io.sbml:'' is not a valid SBML 'SId'.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gap-filling database =  ModelSEED\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "ERROR:cobra.io.sbml:No objective coefficients in model. Unclear what should be optimized\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "#reactions not found in NN-keys:  15 / 101\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Flux through biomass reaction is 1.00000000\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  80213 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  40106 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  20053 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  10026 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  5013 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  2506 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1253 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  626 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  313 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  156 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  78 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  39 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  19 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  9 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  5 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  3 \n",
            "\n",
            "\n",
            "Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored\n",
            "\n",
            "\n",
            " condition is currently:  1 \n",
            "\n",
            "\n",
            "Objective value is 0.000000.\n",
            "Read LP format model from file /tmp/tmp6_2o4xca.lp\n",
            "Reading time = 0.00 seconds\n",
            ": 360 rows, 596 columns, 2694 nonzeros\n",
            "NN gapfilling added 197 new reactions\n",
            "The NN gapfilled model, comes with 298 reactions and 360 metabolites\n"
          ]
        }
      ],
      "source": [
        "!mkdir methanol_draft_models\n",
        "!mkdir methanol_gapfilled_models\n",
        "\n",
        "\n",
        "genomes = os.listdir(\"metabolic_toy_model/files/bins\")\n",
        "\n",
        "for genome in genomes:\n",
        "  model_id = genome.replace(\".fa\", \"\")\n",
        "  input_fasta = \"metabolic_toy_model/files/bins/\" + genome\n",
        "  output_model_sbml = \"methanol_draft_models/\" + model_id + \"_draft.sbml\"\n",
        "  draftModel = reconstruct_draft_model(model_id, input_fasta, output_model_sbml, is_dna_fasta=True)\n",
        "  gapfilledModel = Gapfill(draftModel = output_model_sbml)\n",
        "  cobra.io.write_sbml_model(gapfilledModel.gapfilledModel, filename = \"methanol_gapfilled_models/\" + model_id + \"_gapfilled.sbml\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ae0iC38SuW7x",
        "outputId": "4c46ed94-8c70-4148-a04d-80a9d0a859c9"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "total 1272\n",
            "drwxr-xr-x 2 root root   4096 Mar 13 08:49 .\n",
            "drwxr-xr-x 1 root root   4096 Mar 13 08:32 ..\n",
            "-rw-r--r-- 1 root root 269639 Mar 13 08:49 bin.1_draft.sbml\n",
            "-rw-r--r-- 1 root root 283858 Mar 13 08:46 bin.2_draft.sbml\n",
            "-rw-r--r-- 1 root root 734334 Mar 13 08:47 bin.3_draft.sbml\n",
            "total 6400\n",
            "drwxr-xr-x 2 root root    4096 Mar 13 08:51 .\n",
            "drwxr-xr-x 1 root root    4096 Mar 13 08:32 ..\n",
            "-rw-r--r-- 1 root root 1908407 Mar 13 08:51 bin.1_gapfilled.sbml\n",
            "-rw-r--r-- 1 root root 1990524 Mar 13 08:47 bin.2_gapfilled.sbml\n",
            "-rw-r--r-- 1 root root 2642104 Mar 13 08:49 bin.3_gapfilled.sbml\n"
          ]
        }
      ],
      "source": [
        "!ls -la methanol_draft_models\n",
        "!ls -la methanol_gapfilled_models"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EyT3fJcgm467"
      },
      "source": [
        "### **Now we have the models, what can we do with them?**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5dFL6fjhfv_n"
      },
      "source": [
        "#### **Gut colonization**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "16CRNijDcZdz"
      },
      "source": [
        "#### **Environment ball**\n",
        "\n",
        "Draw environments from a [Dirichlet distribution](https://en.wikipedia.org/wiki/Dirichlet_distribution)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "eFIVDr-Hej-L"
      },
      "outputs": [],
      "source": [
        "def generate_environment_ball(simul_n, transporters, outdir):\n",
        "    '''\n",
        "    simul_n: the number of random environments to generate.\n",
        "    transporters: file containing list of trasporters (with modelSeed ids)\n",
        "    outdir: directory to store results\n",
        "    '''\n",
        "\n",
        "\n",
        "    with open(transporters) as f:\n",
        "        transporters=f.readline().strip().split('\\t')\n",
        "        names = f.readline().strip().split('\\t')\n",
        "    water_index=transporters.index('EX_cpd00001_e')\n",
        "\n",
        "    oxygen_index=transporters.index('EX_cpd00007_e')\n",
        "\n",
        "\n",
        "    random_environments=np.zeros((simul_n,len(transporters)))\n",
        "\n",
        "\n",
        "\n",
        "    for i in range(simul_n):\n",
        "\n",
        "        #Draw from a Dirichlet distribution\n",
        "        a=np.random.dirichlet(np.ones(len(transporters)))\n",
        "\n",
        "        #Set water to a constant value\n",
        "        a[water_index]=1.0\n",
        "\n",
        "        #Set a 50% chance of being anoxic\n",
        "        if np.random.uniform()<0.5:\n",
        "            a[oxygen_index]=0.0\n",
        "        else:\n",
        "            a[oxygen_index]=1.0\n",
        "\n",
        "        #stretch the distribution\n",
        "        a=a*1000\n",
        "        random_environments[i]=a\n",
        "\n",
        "    with open(os.path.join(outdir,   'env_ball_'+str(simul_n)+'.tsv'),'w') as f:\n",
        "\n",
        "        f.write('\\t'.join(transporters) + '\\n')\n",
        "        f.write('\\t'.join(names) + '\\n')\n",
        "\n",
        "\n",
        "        for i in random_environments:\n",
        "\n",
        "            for num in i:\n",
        "                f.write('%.20f\\t'%num)\n",
        "            f.write('\\n')\n",
        "\n",
        "#Example of usage:\n",
        "#main(1000, pathToFileWithListOfTransporters, pathToWriteOutput)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Puk0-IVhlBx7"
      },
      "source": [
        "#### **Methanol**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "![figure](files/figs/garza2020.png)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "include_colab_link": true,
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
